<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>0 - Quality control of sequencing data • miso</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="0 - Quality control of sequencing data">
<meta property="og:description" content="miso">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">miso</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.51.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/00_qa.html">0 - Quality control of sequencing data</a>
    </li>
    <li>
      <a href="../articles/01_preprocessing.html">1 - Preprocessing of sequencing data</a>
    </li>
    <li>
      <a href="../articles/02_amplicon_sequencing.html">2 - Amplicon sequencing analysis</a>
    </li>
    <li>
      <a href="../articles/03_counting.html">3 - Reference counting</a>
    </li>
    <li>
      <a href="../articles/04_sra.html">4 - Submitting data to SRA</a>
    </li>
    <li>
      <a href="../articles/05_plate_layouts.html">5 - Plate Layouts</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dienerlab/miso/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>0 - Quality control of sequencing data</h1>
                        <h4 data-toc-skip class="author">Christian
Diener</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dienerlab/miso/blob/HEAD/vignettes/00_qa.Rmd" class="external-link"><code>vignettes/00_qa.Rmd</code></a></small>
      <div class="hidden name"><code>00_qa.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="quality-assessment-of-you-raw-data">Quality assessment of you raw data<a class="anchor" aria-label="anchor" href="#quality-assessment-of-you-raw-data"></a>
</h2>
<p>One of the first steps after sequencing is usually assessing the
overall quality of your sequencing experiment. Most tools specialize on
a per-sample workflow for that which has not worked well for me since we
usually analyze medium to large batches of samples. The functions here
are meant to facilitate this particular workflow.</p>
<p>First start by loading miso:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dienerlab/miso" class="external-link">miso</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Also loading:</span></span></code></pre>
<pre><code><span><span class="co">##   - dada2=1.28.0</span></span>
<span><span class="co">##   - data.table=1.15.2</span></span>
<span><span class="co">##   - ggplot2=3.5.0</span></span>
<span><span class="co">##   - magrittr=2.0.3</span></span>
<span><span class="co">##   - phyloseq=1.44.0</span></span>
<span><span class="co">##   - ShortRead=1.58.0</span></span>
<span><span class="co">##   - yaml=2.3.8</span></span></code></pre>
<pre><code><span><span class="co">## Found tools:</span></span></code></pre>
<pre><code><span><span class="co">##   - minimap2=2.27-r1193</span></span>
<span><span class="co">##   - samtools=1.19.2</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'miso'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked _by_ 'package:BiocGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     normalize</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:graphics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     layout</span></span></code></pre>
<div class="section level3">
<h3 id="finding-your-files">Finding your files<a class="anchor" aria-label="anchor" href="#finding-your-files"></a>
</h3>
<p><code>miso</code> includes helpers to find your sequencing files. It
works out of the box for normal Illumina file names (the ones you get
from basespace). There is a small example data set in our package which
we can use to illustrate that.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/16S"</span>, package <span class="op">=</span> <span class="st">"miso"</span><span class="op">)</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_read_files.html">find_read_files</a></span><span class="op">(</span><span class="va">path</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">files</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Key: &lt;id&gt;</span></span>
<span><span class="co">##                                                                                      forward</span></span>
<span><span class="co">##                                                                                       &lt;char&gt;</span></span>
<span><span class="co">## 1: /tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/16S/F3D0_S188_L001_R1_001.fastq.gz</span></span>
<span><span class="co">## 2: /tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/16S/F3D1_S189_L001_R1_001.fastq.gz</span></span>
<span><span class="co">## 3: /tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/16S/F3D2_S190_L001_R1_001.fastq.gz</span></span>
<span><span class="co">## 4: /tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/16S/F3D3_S191_L001_R1_001.fastq.gz</span></span>
<span><span class="co">## 5: /tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/16S/Mock_S280_L001_R1_001.fastq.gz</span></span>
<span><span class="co">##                                                                                      reverse</span></span>
<span><span class="co">##                                                                                       &lt;char&gt;</span></span>
<span><span class="co">## 1: /tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/16S/F3D0_S188_L001_R2_001.fastq.gz</span></span>
<span><span class="co">## 2: /tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/16S/F3D1_S189_L001_R2_001.fastq.gz</span></span>
<span><span class="co">## 3: /tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/16S/F3D2_S190_L001_R2_001.fastq.gz</span></span>
<span><span class="co">## 4: /tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/16S/F3D3_S191_L001_R2_001.fastq.gz</span></span>
<span><span class="co">## 5: /tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/16S/Mock_S280_L001_R2_001.fastq.gz</span></span>
<span><span class="co">##        id injection_order  lane</span></span>
<span><span class="co">##    &lt;char&gt;           &lt;num&gt; &lt;num&gt;</span></span>
<span><span class="co">## 1:   F3D0             188     1</span></span>
<span><span class="co">## 2:   F3D1             189     1</span></span>
<span><span class="co">## 3:   F3D2             190     1</span></span>
<span><span class="co">## 4:   F3D3             191     1</span></span>
<span><span class="co">## 5:   Mock             280     1</span></span></code></pre>
<p>As you can see this will try to infer the id and read directions.</p>
</div>
<div class="section level3">
<h3 id="running-quality-assesment">Running Quality assesment<a class="anchor" aria-label="anchor" href="#running-quality-assesment"></a>
</h3>
<p>Now we can run the quality assessment workflow step. This step does
not take a configuration since it does not require that much prior
information. It does take additional arguments to specify the maximum
number of sampled reads and a quality cutoff that you consider
sufficient. The default is a quality score of 10 which is the average
for nanopore reads for instance. This corresponds to a per-base error of
about 10%. We can be a bit more stringent here and use a cutoff of 1%
error (quality score 20).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">quals</span> <span class="op">&lt;-</span> <span class="va">files</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/quality_control.html">quality_control</a></span><span class="op">(</span>min_score <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## INFO [2024-04-29 09:38:26] Running quality assay for forward reads from 5 files.</span></span>
<span><span class="co">## INFO [2024-04-29 09:38:29] Median per base error is 0.016% (median score = 38.00).</span></span>
<span><span class="co">## INFO [2024-04-29 09:38:29] Mean per cycle entropy is 0.808 (in [0, 2]).</span></span>
<span><span class="co">## INFO [2024-04-29 09:38:29] On average we have 8963.80 +- 6059.73 reads per file.</span></span>
<span><span class="co">## INFO [2024-04-29 09:38:29] Running quality assay for reverse reads from 5 files.</span></span>
<span><span class="co">## INFO [2024-04-29 09:38:30] Median per base error is 0.020% (median score = 37.00).</span></span>
<span><span class="co">## INFO [2024-04-29 09:38:30] Mean per cycle entropy is 0.731 (in [0, 2]).</span></span>
<span><span class="co">## INFO [2024-04-29 09:38:30] On average we have 8963.80 +- 6059.73 reads per file.</span></span></code></pre>
<p>This runs in parallel and will report global quality metrics on the
logging interface. We see that we have relatively shallow sequencing
data, but the average quality and entropy is pretty good.</p>
<div class="section level4">
<h4 id="quality-plots">Quality plots<a class="anchor" aria-label="anchor" href="#quality-plots"></a>
</h4>
<p>In order to indentify points where the quality tanks we can have a
look at the average quality profile.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">quals</span><span class="op">$</span><span class="va">quality_plot</span></span></code></pre></div>
<p><img src="00_qa_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>This is a quick global plot giving us the overall average quality per
sample. The 5’ (first 10bp) qualities are usually artificially lowered
by Illumina sequencers and have some lower confidence. We also see some
declining qualities on the 3’ of the reverse reads. This is all pretty
typical for MiSeq data.</p>
</div>
<div class="section level4">
<h4 id="length-plots">Length plots<a class="anchor" aria-label="anchor" href="#length-plots"></a>
</h4>
<p>One additional questions is if we have samples were reads are shorter
of have strongly declining qualities compared to other samples. This can
be evaluated by the length plot.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">quals</span><span class="op">$</span><span class="va">length_plot</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 20 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">## (`geom_raster()`).</span></span></code></pre>
<p><img src="00_qa_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>You usually want to see horizontal bands of a single color. If the
colors become lighter towards the right that means that some samples
have shorter reads after trimming which could complicate some
workflows.</p>
</div>
<div class="section level4">
<h4 id="entropy-plots">Entropy plots<a class="anchor" aria-label="anchor" href="#entropy-plots"></a>
</h4>
<p>We can also check how heterogenous the reads positions are. This is
quantified by the entropy which tells us how varying a single position
is. Any value appreciably larger than 0 is usually fine.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">quals</span><span class="op">$</span><span class="va">entropy_plot</span></span></code></pre></div>
<p><img src="00_qa_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>16S amplified regions tend to have some lower variability on the ends
which is what we see by the lower entropy on the beginning of the
forward and reverse reads (the ends of the amplicon).</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Christian Diener.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
