<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>2 - Amplicon sequencing analysis • miso</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="2 - Amplicon sequencing analysis">
<meta property="og:description" content="miso">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">miso</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.50.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/00_qa.html">0 - Quality control of sequencing data</a>
    </li>
    <li>
      <a href="../articles/01_preprocessing.html">1 - Preprocessing of sequencing data</a>
    </li>
    <li>
      <a href="../articles/02_amplicon_sequencing.html">2 - Amplicon sequencing analysis</a>
    </li>
    <li>
      <a href="../articles/03_counting.html">3 - Reference counting</a>
    </li>
    <li>
      <a href="../articles/04_sra.html">4 - Submitting data to SRA</a>
    </li>
    <li>
      <a href="../articles/05_plate_layouts.html">5 - Plate Layouts</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dienerlab/miso/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>2 - Amplicon sequencing analysis</h1>
                        <h4 data-toc-skip class="author">Christian
Diener</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dienerlab/miso/blob/HEAD/vignettes/02_amplicon_sequencing.Rmd" class="external-link"><code>vignettes/02_amplicon_sequencing.Rmd</code></a></small>
      <div class="hidden name"><code>02_amplicon_sequencing.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="amplicon-sequencing-data-analysis">Amplicon sequencing data analysis<a class="anchor" aria-label="anchor" href="#amplicon-sequencing-data-analysis"></a>
</h2>
<p>We can combine what we have seen in the quality assessment and
preprocessing tutoarials with specific workflow steps for amplicon
sequencing, for instance 16S rRNA gene sequencing.</p>
<p>As before we will load <code>miso</code> and use our example
data:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dienerlab/miso" class="external-link">miso</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Also loading:</span></span></code></pre>
<pre><code><span><span class="co">##   - dada2=1.28.0</span></span>
<span><span class="co">##   - data.table=1.15.2</span></span>
<span><span class="co">##   - ggplot2=3.5.0</span></span>
<span><span class="co">##   - magrittr=2.0.3</span></span>
<span><span class="co">##   - phyloseq=1.44.0</span></span>
<span><span class="co">##   - ShortRead=1.58.0</span></span>
<span><span class="co">##   - yaml=2.3.8</span></span></code></pre>
<pre><code><span><span class="co">## Found tools:</span></span></code></pre>
<pre><code><span><span class="co">##   - minimap2=2.27-r1193</span></span>
<span><span class="co">##   - samtools=1.19.2</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'miso'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked _by_ 'package:BiocGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     normalize</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:graphics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     layout</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/16S"</span>, package <span class="op">=</span> <span class="st">"miso"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="quality-assessment">Quality assessment<a class="anchor" aria-label="anchor" href="#quality-assessment"></a>
</h3>
<p>As before a good first step is to look at the qualities of the raw
data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">quals</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_read_files.html">find_read_files</a></span><span class="op">(</span><span class="va">path</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/quality_control.html">quality_control</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## INFO [2024-04-25 12:55:47] Running quality assay for forward reads from 5 files.</span></span>
<span><span class="co">## INFO [2024-04-25 12:55:50] Median per base error is 0.016% (median score = 38.00).</span></span>
<span><span class="co">## INFO [2024-04-25 12:55:50] Mean per cycle entropy is 0.808 (in [0, 2]).</span></span>
<span><span class="co">## INFO [2024-04-25 12:55:50] On average we have 8963.80 +- 6059.73 reads per file.</span></span>
<span><span class="co">## INFO [2024-04-25 12:55:50] Running quality assay for reverse reads from 5 files.</span></span>
<span><span class="co">## INFO [2024-04-25 12:55:51] Median per base error is 0.020% (median score = 37.00).</span></span>
<span><span class="co">## INFO [2024-04-25 12:55:51] Mean per cycle entropy is 0.731 (in [0, 2]).</span></span>
<span><span class="co">## INFO [2024-04-25 12:55:51] On average we have 8963.80 +- 6059.73 reads per file.</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">quals</span><span class="op">$</span><span class="va">quality_plot</span></span></code></pre></div>
<p><img src="02_amplicon_sequencing_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<p>Again we see that we might want to truncate the reads a bit on the 3’
ends to avoid the dip in quality. Depending on your amplified fragment
make sure that this leaves sufficient overlap for merging (requires
&gt;20bp).</p>
</div>
<div class="section level3">
<h3 id="denoise">Denoise<a class="anchor" aria-label="anchor" href="#denoise"></a>
</h3>
<p>We can now perform our preprocessing and denoise workflow to obtain
the amplicon sequence variants (ASVs) in our samples. We can chain our
preprocessing and denoise workflows from our original quality
assessment. For clarity we will define the confguration for the analysis
on top.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">config</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    preprocess <span class="op">=</span> <span class="fu"><a href="../reference/config_preprocess.html">config_preprocess</a></span><span class="op">(</span></span>
<span>        trimLeft <span class="op">=</span> <span class="fl">10</span>,  <span class="co"># this is the default</span></span>
<span>        trunLen <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">240</span>, <span class="fl">180</span><span class="op">)</span>,  <span class="co"># forward and reverse truncation</span></span>
<span>        out_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>  <span class="co"># will store filtered files in a temporary dir</span></span>
<span>    <span class="op">)</span>,</span>
<span>    denoise <span class="op">=</span> <span class="fu"><a href="../reference/config_denoise.html">config_denoise</a></span><span class="op">(</span><span class="op">)</span>  <span class="co"># will only use defaults</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">denoised</span> <span class="op">&lt;-</span> <span class="va">quals</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/preprocess.html">preprocess</a></span><span class="op">(</span><span class="va">config</span><span class="op">$</span><span class="va">preprocess</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/denoise.html">denoise</a></span><span class="op">(</span><span class="va">config</span><span class="op">$</span><span class="va">denoise</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## INFO [2024-04-25 12:55:52] Preprocessing reads for 5 paired-end samples...</span></span>
<span><span class="co">## INFO [2024-04-25 12:55:58] 3.23e+04/4.48e+04 (73.36%) reads passed preprocessing.</span></span>
<span><span class="co">## INFO [2024-04-25 12:55:58] Running DADA2 on 1 run(s) from a sample of up to 2.5e+08 bases.</span></span>
<span><span class="co">## INFO [2024-04-25 12:55:58] Learning errors for run `all` (5 samples)...</span></span>
<span><span class="co">## INFO [2024-04-25 12:56:53] Dereplicating run `all` (5 samples)...</span></span>
<span><span class="co">## INFO [2024-04-25 12:56:54] Inferring sequence variants for run `all`...</span></span>
<span><span class="co">## INFO [2024-04-25 12:57:08] Found 164 sequence variants in run `all`...</span></span>
<span><span class="co">## INFO [2024-04-25 12:57:08] Removed 5/164 sequence variants as chimeric from run all (1.21% of reads)</span></span>
<span><span class="co">## INFO [2024-04-25 12:57:08] Finished run `all`.</span></span>
<span><span class="co">## INFO [2024-04-25 12:57:08] Merged sequence tables. Found a total of 159 ASVs. Assigning taxonomy...</span></span>
<span><span class="co">## INFO [2024-04-25 12:57:08] Downloading taxa db to /tmp/Rtmp959rH9/taxa.fna.gz...</span></span>
<span><span class="co">## INFO [2024-04-25 12:59:43] Downloading species db to /tmp/Rtmp959rH9/species.fna.gz...</span></span>
<span><span class="co">## INFO [2024-04-25 13:00:13] Reads with taxonomic classification: Kingdom=100%, Phylum=100%, Class=100%, Order=100%, Family=96.4%, Genus=53.6%, Species=18.8%, Species=18.8%</span></span>
<span><span class="co">## INFO [2024-04-25 13:00:13] Hashing 159 sequence variants.</span></span></code></pre>
<p>This will run both workflow steps sequentially and will use all
available CPUs by default. You will get some disgnostic output on the
logging interface but you can also inspect those in the returned
artifact.</p>
<p>For instance to see how many reads were conserved in each processing
step:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">denoised</span><span class="op">$</span><span class="va">passed_reads</span></span></code></pre></div>
<pre><code><span><span class="co">##      raw preprocessed     id derep_forward derep_reverse denoised_forward</span></span>
<span><span class="co">##    &lt;num&gt;        &lt;num&gt; &lt;char&gt;         &lt;int&gt;         &lt;int&gt;            &lt;int&gt;</span></span>
<span><span class="co">## 1:  7793         5616   F3D0          5616          5616             5461</span></span>
<span><span class="co">## 2:  5869         4388   F3D1          4388          4388             4280</span></span>
<span><span class="co">## 3: 19620        13787   F3D2         13787         13787            13633</span></span>
<span><span class="co">## 4:  6758         4527   F3D3          4527          4527             4423</span></span>
<span><span class="co">## 5:  4779         3953   Mock          3953          3953             3933</span></span>
<span><span class="co">##    denoised_reverse merged non_chimera</span></span>
<span><span class="co">##               &lt;int&gt;  &lt;int&gt;       &lt;num&gt;</span></span>
<span><span class="co">## 1:             5415   4795        4795</span></span>
<span><span class="co">## 2:             4255   4005        4005</span></span>
<span><span class="co">## 3:            13583  12872       12700</span></span>
<span><span class="co">## 4:             4328   3760        3578</span></span>
<span><span class="co">## 5:             3934   3895        3895</span></span></code></pre>
<p>We see that most samples were conserved (as reported in the
logging).</p>
<p>The abundance of each sequence variant is saved in
<code>denoised$feature_tables</code> which contains abundances for
samples x ASVs. The ASVs here have been named by their md5 hash value
and you can get the taxonomy and original sequence from
<code>denoised$taxonomy</code>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">denoised</span><span class="op">$</span><span class="va">taxonomy</span><span class="op">[</span><span class="fl">1</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                                  Kingdom    Phylum         Class        </span></span>
<span><span class="co">## c69a3db92e1dcdebb01dfaa5b5795559 "Bacteria" "Bacteroidota" "Bacteroidia"</span></span>
<span><span class="co">##                                  Order           Family           Genus Species</span></span>
<span><span class="co">## c69a3db92e1dcdebb01dfaa5b5795559 "Bacteroidales" "Muribaculaceae" NA    NA     </span></span>
<span><span class="co">##                                  Species</span></span>
<span><span class="co">## c69a3db92e1dcdebb01dfaa5b5795559 NA     </span></span>
<span><span class="co">##                                  sequence                                                                                                                                                                                                                                                    </span></span>
<span><span class="co">## c69a3db92e1dcdebb01dfaa5b5795559 "ACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGCAGGCGGAAGATCAAGTCAGCGGTAAAATTGAGAGGCTCAACCTCTTCGAGCCGTTGAAACTGGTTTTCTTGAGTGAGCGAGAAGTATGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACTCCGATTGCGAAGGCAGCATACCGGCGCTCAACTGACGCTCATGCACGAAAGTGTGGGTATCGAACAG"</span></span></code></pre>
<p>To see all steps run to obtain that output you can check the
provenance:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">denoised</span><span class="op">$</span><span class="va">steps</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "quality_control" "preprocess"      "denoise"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="managing-several-sequencing-runs">Managing several sequencing runs<a class="anchor" aria-label="anchor" href="#managing-several-sequencing-runs"></a>
</h3>
<p>The <code>denoise</code> workflow step supports multiple sequencing
runs in the same experiment. This requires that your files data table
has a “run” column. If you sequencing files are organized such that
every subfolder is a run you can simply use
<code>find_read_files(path, dirs_are_runs = TRUE)</code>. Otherwise, you
have to specify the run by hand. We will do so for our files here:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">files</span> <span class="op">&lt;-</span> <span class="va">quals</span><span class="op">$</span><span class="va">files</span></span>
<span><span class="va">files</span><span class="op">[</span>, <span class="va">run</span> <span class="op">:=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, times <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">files</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Key: &lt;id&gt;</span></span>
<span><span class="co">##                                                                                       forward</span></span>
<span><span class="co">##                                                                                        &lt;char&gt;</span></span>
<span><span class="co">## 1: /tmp/Rtmp0LsDXk/temp_libpath20758c3f036fbb/miso/extdata/16S/F3D0_S188_L001_R1_001.fastq.gz</span></span>
<span><span class="co">## 2: /tmp/Rtmp0LsDXk/temp_libpath20758c3f036fbb/miso/extdata/16S/F3D1_S189_L001_R1_001.fastq.gz</span></span>
<span><span class="co">## 3: /tmp/Rtmp0LsDXk/temp_libpath20758c3f036fbb/miso/extdata/16S/F3D2_S190_L001_R1_001.fastq.gz</span></span>
<span><span class="co">## 4: /tmp/Rtmp0LsDXk/temp_libpath20758c3f036fbb/miso/extdata/16S/F3D3_S191_L001_R1_001.fastq.gz</span></span>
<span><span class="co">## 5: /tmp/Rtmp0LsDXk/temp_libpath20758c3f036fbb/miso/extdata/16S/Mock_S280_L001_R1_001.fastq.gz</span></span>
<span><span class="co">##                                                                                       reverse</span></span>
<span><span class="co">##                                                                                        &lt;char&gt;</span></span>
<span><span class="co">## 1: /tmp/Rtmp0LsDXk/temp_libpath20758c3f036fbb/miso/extdata/16S/F3D0_S188_L001_R2_001.fastq.gz</span></span>
<span><span class="co">## 2: /tmp/Rtmp0LsDXk/temp_libpath20758c3f036fbb/miso/extdata/16S/F3D1_S189_L001_R2_001.fastq.gz</span></span>
<span><span class="co">## 3: /tmp/Rtmp0LsDXk/temp_libpath20758c3f036fbb/miso/extdata/16S/F3D2_S190_L001_R2_001.fastq.gz</span></span>
<span><span class="co">## 4: /tmp/Rtmp0LsDXk/temp_libpath20758c3f036fbb/miso/extdata/16S/F3D3_S191_L001_R2_001.fastq.gz</span></span>
<span><span class="co">## 5: /tmp/Rtmp0LsDXk/temp_libpath20758c3f036fbb/miso/extdata/16S/Mock_S280_L001_R2_001.fastq.gz</span></span>
<span><span class="co">##        id injection_order  lane   run</span></span>
<span><span class="co">##    &lt;char&gt;           &lt;num&gt; &lt;num&gt; &lt;int&gt;</span></span>
<span><span class="co">## 1:   F3D0             188     1     1</span></span>
<span><span class="co">## 2:   F3D1             189     1     1</span></span>
<span><span class="co">## 3:   F3D2             190     1     2</span></span>
<span><span class="co">## 4:   F3D3             191     1     2</span></span>
<span><span class="co">## 5:   Mock             280     1     3</span></span></code></pre>
<p>Rerunning the previous workflow with this file list will now give a
slightly different output:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">byrun</span> <span class="op">&lt;-</span> <span class="va">files</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/preprocess.html">preprocess</a></span><span class="op">(</span><span class="va">config</span><span class="op">$</span><span class="va">preprocess</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>         <span class="fu"><a href="../reference/denoise.html">denoise</a></span><span class="op">(</span><span class="va">config</span><span class="op">$</span><span class="va">denoise</span><span class="op">)</span></span></code></pre></div>
<p>Error profiles will now be estimated for each run individually and
final tables will be merged at the end. This is slower than having all
files in one run so you should always try to distribute your samples
across as few sequencing runsps as possible.</p>
</div>
<div class="section level3">
<h3 id="converting-to-phyloseq">Converting to phyloseq<a class="anchor" aria-label="anchor" href="#converting-to-phyloseq"></a>
</h3>
<p>For all further analyses you might want to work with a data type
which is more suited for amplicon abundance data. You can directly
obtain a phyloseq object with</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as_phyloseq.html">as_phyloseq</a></span><span class="op">(</span><span class="va">denoised</span><span class="op">)</span></span></code></pre></div>
<p>It is possible to pass in metadata as a data frame with
<code>to_phyloseq(object, metadata = df)</code>.</p>
<p>You can now use all additional functions from <code>miso</code> on
that. For instance we can get the ASV counts in a tidy data table:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">asv_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taxa_count.html">taxa_count</a></span><span class="op">(</span><span class="va">ps</span>, lev <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span>
<span><span class="va">asv_counts</span><span class="op">[</span><span class="va">sample</span> <span class="op">==</span> <span class="st">"Mock"</span> <span class="op">&amp;</span> <span class="va">reads</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##     sample                             taxa reads</span></span>
<span><span class="co">##     &lt;char&gt;                           &lt;fctr&gt; &lt;int&gt;</span></span>
<span><span class="co">##  1:   Mock e941355a23a624aef3fe3e4caf9bc301   564</span></span>
<span><span class="co">##  2:   Mock 532c8d14997a65c94cdabbf41c8f2647    90</span></span>
<span><span class="co">##  3:   Mock 35b4a92774d4fe5a2d33367602b223a8   378</span></span>
<span><span class="co">##  4:   Mock b2df6175ad0400ad1301072c5d64a75f   348</span></span>
<span><span class="co">##  5:   Mock cc36ce261772980980931b51d3d9a379   309</span></span>
<span><span class="co">##  6:   Mock 7b106324ebe9486ccd49abc287a6de45   288</span></span>
<span><span class="co">##  7:   Mock ed8390cd975c833a487e2678b248bf13   272</span></span>
<span><span class="co">##  8:   Mock 454886158d7f3bc1b770132e75a1c87f   243</span></span>
<span><span class="co">##  9:   Mock ea98650694b6343f2ab88943711c2595   206</span></span>
<span><span class="co">## 10:   Mock 5ff1477099ca90cc5c48b7b231b735eb   203</span></span>
<span><span class="co">## 11:   Mock 75d2399d0d5f38dcd98e7f4c91271272   173</span></span>
<span><span class="co">## 12:   Mock bf4c86674249ac9c02f5701d3dfdfd31   163</span></span>
<span><span class="co">## 13:   Mock 8a214389774631998cca74c5b0088d8b   148</span></span>
<span><span class="co">## 14:   Mock 37abc7707b7be944fb565da5297d3175   136</span></span>
<span><span class="co">## 15:   Mock f43b31caaa0047cf889f30ec8e56ec17   131</span></span>
<span><span class="co">## 16:   Mock 664acad52badcb22b536dd609d7c4f24   111</span></span>
<span><span class="co">## 17:   Mock e0c48dcc0eb1e70ac90edb8107d4e0d4    69</span></span>
<span><span class="co">## 18:   Mock b6a0a5f9e9187f987e31ebf1c8c4bd1d    48</span></span>
<span><span class="co">## 19:   Mock b6f891b70ee28c80ed84667ea428baf3    15</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                   species</span></span>
<span><span class="co">##                                                                                                                                                                                                                                                                                    &lt;char&gt;</span></span>
<span><span class="co">##  1:              Staphylococcus CGTAGGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGCGCGTAGGCGGTTTTTTAAGTCTGATGTGAAAGCCCACGGCTCAACCGTGGAGGGTCATTGGAAACTGGAAAACTTGAGTGCAGAAGAGGAAAGTGGAATTCCATGTGTAGCGGTGAAATGCGCAGAGATATGGAGGAACACCAGTGGCGAAGGCGACTTTCTGGTCTGTAACTGACGCTGATGTGCGAAAGCGTGGGGATCAAACA</span></span>
<span><span class="co">##  2:               Lactobacillus CGTAGGTGGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGTGCAGGCGGTTCAATAAGTCTGATGTGAAAGCCTTCGGCTCAACCGGAGAATTGCATCAGAAACTGTTGAACTTGAGTGCAGAAGAGGAGAGTGGAACTCCATGTGTAGCGGTGGAATGCGTAGATATATGGAAGAACACCAGTGGCGAAGGCGGCTCTCTGGTCTGCAACTGACGCTGAGGCTCGAAAGCATGGGTAGCGAACA</span></span>
<span><span class="co">##  3:                  Acinetobacter AGGGTGCGAGCGTTAATCGGATTTACTGGGCGTAAAGCGTGCGTAGGCGGCTTATTAAGTCGGATGTGAAATCCCCGAGCTTAACTTGGGAATTGCATTCGATACTGGTGAGCTAGAGTATGGGAGAGGATGGTAGAATTCCAGGTGTAGCGGTGAAATGCGTAGAGATCTGGAGGAATACCGATGGCGAAGGCAGCCATCTGGCCTAATACTGACGCTGAGGTACGAAAGCATGGGGAGCAAACA</span></span>
<span><span class="co">##  4:                    Bacillus CGTAGGTGGCAAGCGTTATCCGGAATTATTGGGCGTAAAGCGCGCGCAGGTGGTTTCTTAAGTCTGATGTGAAAGCCCACGGCTCAACCGTGGAGGGTCATTGGAAACTGGGAGACTTGAGTGCAGAAGAGGAAAGTGGAATTCCATGTGTAGCGGTGAAATGCGTAGAGATATGGAGGAACACCAGTGGCGAAGGCGACTTTCTGGTCTGTAACTGACACTGAGGCGCGAAAGCGTGGGGAGCAAACA</span></span>
<span><span class="co">##  5:                   Helicobacter GAGGGTGCAAGCGTTACTCGGAATCACTGGGCGTAAAGAGCGCGTAGGCGGGATAGTCAGTCAGGTGTGAAATCCTATGGCTTAACCATAGAACTGCATTTGAAACTACTATTCTAGAGTGTGGGAGAGGTAGGTGGAATTCTTGGTGTAGGGGTAAAATCCGTAGAGATCAAGAGGAATACTCATTGCGAAGGCGACCTGCTGGAACATTACTGACGCTGATTGCGCGAAAGCGTGGGGAGCAAA</span></span>
<span><span class="co">##  6: Clostridium sensu stricto 1 CGTAGGTGGCAAGCGTTGTCCGGATTTACTGGGCGTAAAGGGAGCGTAGGTGGATATTTAAGTGGGATGTGAAATACTCGGGCTTAACCTGGGTGCTGCATTCCAAACTGGATATCTAGAGTGCAGGAGAGGAAAGTAGAATTCCTAGTGTAGCGGTGAAATGCGTAGAGATTAGGAAGAATACCAGTGGCGAAGGCGACTTTCTGGACTGTAACTGACACTGAGGCTCGAAAGCGTGGGGAGCAAACA</span></span>
<span><span class="co">##  7:                   Neisseria CGTAGGGTGCGAGCGTTAATCGGAATTACTGGGCGTAAAGCGGGCGCAGACGGTTACTTAAGCAGGATGTGAAATCCCCGGGCTCAACCCGGGAACTGCGTTCTGAACTGGGTGACTCGAGTGTGTCAGAGGGAGGTAGAATTCCACGTGTAGCAGTGAAATGCGTAGAGATGTGGAGGAATACCGATGGCGAAGGCAGCCTCCTGGGACAACACTGACGTTCATGCCCGAAAGCGTGGGTAGCAAACA</span></span>
<span><span class="co">##  8:                 Actinomyces CGTAGGGCGCGAGCGTTGTCCGGAATTATTGGGCGTAAAGGGCTTGTAGGCGGTTGGTCGCGTCTGCCGTGAAATCCTCTGGCTTAACTGGGGGCGTGCGGTGGGTACGGGCTGACTTGAGTGCGGTAGGGGAGACTGGAACTCCTGGTGTAGCGGTGGAATGCGCAGATATCAGGAAGAACACCGGTGGCGAAGGCGGGTCTCTGGGCCGTTACTGACGCTGAGGAGCGAAAGCGTGGGGAGCGAACA</span></span>
<span><span class="co">##  9:                 Bacteroides CGGAGGATCCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGAGCGTAGATGGATGTTTAAGTCAGTTGTGAAAGTTTGCGGCTCAACCGTAAAATTGCAGTTGATACTGGATATCTTGAGTGCAGTTGAGGCAGGCGGAATTCGTGGTGTAGCGGTGAAATGCTTAGATATCACGAAGAACTCCGATTGCGAAGGCAGCCTGCTAAGCTGCAACTGACATTGAGGCTCGAAAGTGTGGGTATCAAACA</span></span>
<span><span class="co">## 10:              Streptococcus ACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGGGAGCGCAGGCGGTCAGGAAAGTCTGGAGTAAAAGGCTATGGCTCAACCATAGTGTGCTCTGGAAACTGTCTGACTTGAGTGCAGAAGGGGAGAGTGGAATTCCATGTGTAGCGGTGAAATGCGTAGATATATGGAGGAACACCAGTGGCGAAAGCGGCTCTCTGGTCTGTCACTGACGCTGAGGCTCGAAAGCGTGGGTAGCGAACAG</span></span>
<span><span class="co">## 11:                Enterococcus CGTAGGTGGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTTCTTAAGTCTGATGTGAAAGCCCCCGGCTCAACCGGGGAGGGTCATTGGAAACTGGGAGACTTGAGTGCAGAAGAGGAGAGTGGAATTCCATGTGTAGCGGTGAAATGCGTAGATATATGGAGGAACACCAGTGGCGAAGGCGGCTCTCTGGTCTGTAACTGACGCTGAGGCTCGAAAGCGTGGGGAGCAAACA</span></span>
<span><span class="co">## 12:        Escherichia-Shigella CGGAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCACGCAGGCGGTTTGTTAAGTCAGATGTGAAATCCCCGGGCTCAACCTGGGAACTGCATCTGATACTGGCAAGCTTGAGTCTCGTAGAGGGGGGTAGAATTCCAGGTGTAGCGGTGAAATGCGTAGAGATCTGGAGGAATACCGGTGGCGAAGGCGGCCCCCTGGACGAAGACTGACGCTCAGGTGCGAAAGCGTGGGGAGCAAACA</span></span>
<span><span class="co">## 13:                    Listeria CGTAGGTGGCAAGCGTTGTCCGGATTTATTGGGCGTAAAGCGCGCGCAGGCGGTCTTTTAAGTCTGATGTGAAAGCCCCCGGCTTAACCGGGGAGGGTCATTGGAAACTGGAAGACTGGAGTGCAGAAGAGGAGAGTGGAATTCCACGTGTAGCGGTGAAATGCGTAGATATGTGGAGGAACACCAGTGGCGAAGGCGACTCTCTGGTCTGTAACTGACGCTGAGGCGCGAAAGCGTGGGGAGCAAACA</span></span>
<span><span class="co">## 14:                 Pseudomonas CGAAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCGCGTAGGTGGTTCAGCAAGTTGGATGTGAAATCCCCGGGCTCAACCTGGGAACTGCATCCAAAACTACTGAGCTAGAGTACGGTAGAGGGTGGTGGAATTTCCTGTGTAGCGGTGAAATGCGTAGATATAGGAAGGAACACCAGTGGCGAAGGCGACCACCTGGACTGATACTGACACTGAGGTGCGAAAGCGTGGGGAGCAAACA</span></span>
<span><span class="co">## 15:              Streptococcus ACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTCTTTAAGTCTGAAGTTAAAGGCAGTGGCTTAACCATTGTACGCTTTGGAAACTGGAGGACTTGAGTGCAGAAGGGGAGAGTGGAATTCCATGTGTAGCGGTGAAATGCGTAGATATATGGAGGAACACCGGTGGCGAAAGCGGCTCTCTGGTCTGTAACTGACGCTGAGGCTCGAAAGCGTGGGGAGCAAACAG</span></span>
<span><span class="co">## 16:                  Deinococcus CGGAGGGTGCAAGCGTTACCCGGAATCACTGGGCGTAAAGGGCGTGTAGGCGGAAATTTAAGTCTGGTTTTAAAGACCGGGGCTCAACCTCGGGGATGGACTGGATACTGGATTTCTTGACCTCTGGAGAGGTAACTGGAATTCCTGGTGTAGCGGTGGAATGCGTAGATACCAGGAGGAACACCAATGGCGAAGGCAAGTTACTGGACAGAAGGTGACGCTGAGGCGCGAAAGTGTGGGGAGCAAAC</span></span>
<span><span class="co">## 17:               Porphyromonas CGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGTAGGTTGTTCGGTAAGTCAGCGGTGAAACCTGAGCGCTCAACGTTCAGCCTGCCGTTGAAACTGCCGGGCTTGAGTTCAGCGGCGGCAGGCGGAATTCGTGGTGTAGCGGTGAAATGCATAGATATCACGAGGAACTCCGATTGCGAAGGCAGCTTGCCATACTGCGACTGACACTGAAGCACGAAGGCGTGGGTATCAAACA</span></span>
<span><span class="co">## 18:                 Rhodobacter CGGAGGGGGCTAGCGTTATTCGGAATTACTGGGCGTAAAGCGCACGTAGGCGGATCGGAAAGTCAGAGGTGAAATCCCAGGGCTCAACCCTGGAACTGCCTTTGAAACTCCCGATCTTGAGGTCGAGAGAGGTGAGTGGAATTCCGAGTGTAGAGGTGAAATTCGTAGATATTCGGAGGAACACCAGTGGCGAAGGCGGCTCACTGGCTCGATACTGACGCTGAGGTGCGAAAGCGTGGGGAGCAAACA</span></span>
<span><span class="co">## 19:              Streptococcus ACGTAGGTCCCGAGCGTTGTCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTGTGGCTTAACCATAGTAGGCTTTGGAAACTGTTTAACTTGAGTGCAAGAGGGGAGAGTGGAATTCCATGTGTAGCGGTGAAATGCGTAGATATATGGAGGAACACCGGTGGCGAAAGCGGCTCTCTGGCTTGTAACTGACGCTGAGGCTCGAAAGCGTGGGGAGCAAACAG</span></span></code></pre>
<p>Here we see that we have identified 10 ASVs in the Mock sample. The
exact number of strains in that sample is 20, which we don’t find since
we don’t have enough reads available to learn a perfect error model.
This is not surprising since those are only a few files from that
particular study.</p>
<p>We could also plot the phyla distribution across the samples:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_taxa.html">plot_taxa</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Invalid .internal.selfref detected and fixed by taking a (shallow)</span></span>
<span><span class="co">## copy of the data.table so that := can add this new column by reference. At an</span></span>
<span><span class="co">## earlier point, this data.table has been copied by R (or was created manually</span></span>
<span><span class="co">## using structure() or similar). Avoid names&lt;- and attr&lt;- which in R currently</span></span>
<span><span class="co">## (and oddly) may copy the whole data.table. Use set* syntax instead to avoid</span></span>
<span><span class="co">## copying: ?set, ?setnames and ?setattr. If this message doesn't help, please</span></span>
<span><span class="co">## report your use case to the data.table issue tracker so the root cause can be</span></span>
<span><span class="co">## fixed or this message improved.</span></span></code></pre>
<p><img src="02_amplicon_sequencing_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Christian Diener.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
