<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>2 - Amplicon sequencing analysis • mbtools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="2 - Amplicon sequencing analysis">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mbtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.40.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/00_qa.html">0 - Quality control of sequencing data</a>
    </li>
    <li>
      <a href="../articles/01_preprocessing.html">1 - Preprocessing of sequencing data</a>
    </li>
    <li>
      <a href="../articles/02_amplicon_sequencing.html">2 - Amplicon sequencing analysis</a>
    </li>
    <li>
      <a href="../articles/03_shotgun_align.html">3 - Shotgun data - alignments</a>
    </li>
    <li>
      <a href="../articles/04_long_align.html">4 - Long read data - alignments</a>
    </li>
    <li>
      <a href="../articles/05_lineage.html">5 - Lineage calling with SLIMM</a>
    </li>
    <li>
      <a href="../articles/06_counting.html">6 - Reference counting</a>
    </li>
    <li>
      <a href="../articles/07_workflow_management.html">7 - Workflow management</a>
    </li>
    <li>
      <a href="../articles/08_sra.html">8 - Submitting data to SRA</a>
    </li>
    <li>
      <a href="../articles/09_plate_layouts.html">9 - Plate Layouts</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Gibbons-Lab/mbtools">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>2 - Amplicon sequencing analysis</h1>
                        <h4 class="author">Christian Diener</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Gibbons-Lab/mbtools/blob/master/vignettes/02_amplicon_sequencing.Rmd"><code>vignettes/02_amplicon_sequencing.Rmd</code></a></small>
      <div class="hidden name"><code>02_amplicon_sequencing.Rmd</code></div>

    </div>

    
    
<div id="amplicon-sequencing-data-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#amplicon-sequencing-data-analysis" class="anchor"></a>Amplicon sequencing data analysis</h1>
<p>We can combine what we have seen in the quality assessment and preprocessing tutoarials with specific workflow steps for amplicon sequencing, for instance 16S rRNA gene sequencing.</p>
<p>As before we will load <code>mbtools</code> and use our example data:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(mbtools)</span></code></pre></div>
<pre><code>## Also loading:</code></pre>
<pre><code>##   - dada2=1.12.1
##   - data.table=1.12.6
##   - ggplot2=3.2.1
##   - magrittr=1.5
##   - phyloseq=1.28.0
##   - ShortRead=1.42.0
##   - yaml=2.2.0</code></pre>
<pre><code>## Found tools:</code></pre>
<pre><code>##   - minimap2=2.17-r941
##   - slimm=0.3.4
##   - samtools=1.9</code></pre>
<pre><code>## 
## Attaching package: 'mbtools'</code></pre>
<pre><code>## The following object is masked _by_ 'package:BiocGenerics':
## 
##     normalize</code></pre>
<pre><code>## The following object is masked from 'package:graphics':
## 
##     layout</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>path &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata/16S"</span>, <span class="dt">package =</span> <span class="st">"mbtools"</span>)</span></code></pre></div>
<div id="quality-assessment" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-assessment" class="anchor"></a>Quality assessment</h2>
<p>As before a good first step is to look at the qualities of the raw data.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>quals &lt;-<span class="st"> </span><span class="kw"><a href="../reference/find_read_files.html">find_read_files</a></span>(path) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/quality_control.html">quality_control</a></span>()</span></code></pre></div>
<pre><code>## INFO [2019-11-05 08:12:09] Running quality assay for forward reads from 5 files.
## INFO [2019-11-05 08:12:13] Median per base error is 0.016% (median score = 38.00).
## INFO [2019-11-05 08:12:13] Mean per cycle entropy is 0.809 (in [0, 2]).
## INFO [2019-11-05 08:12:13] On average we have 8963.80 +- 6059.73 reads per file.
## INFO [2019-11-05 08:12:13] Running quality assay for reverse reads from 5 files.
## INFO [2019-11-05 08:12:16] Median per base error is 0.020% (median score = 37.00).
## INFO [2019-11-05 08:12:16] Mean per cycle entropy is 0.731 (in [0, 2]).
## INFO [2019-11-05 08:12:16] On average we have 8963.80 +- 6059.73 reads per file.</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a>quals<span class="op">$</span>quality_plot</span></code></pre></div>
<p><img src="02_amplicon_sequencing_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<p>Again we see that we might want to truncate the reads a bit on the 3’ ends to avoid the dip in quality. Depending on your amplified fragment make sure that this leaves sufficient overlap for merging (requires &gt;20bp).</p>
</div>
<div id="denoise" class="section level2">
<h2 class="hasAnchor">
<a href="#denoise" class="anchor"></a>Denoise</h2>
<p>We can now perform our preprocessing and denoise workflow to obtain the amplicon sequence variants (ASVs) in our samples. We can chain our preprocessing and denoise workflows from our original quality assessment. For clarity we will define the confguration for the analysis on top.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>config &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(</span>
<span id="cb13-2"><a href="#cb13-2"></a>    <span class="dt">preprocess =</span> <span class="kw"><a href="../reference/config_preprocess.html">config_preprocess</a></span>(</span>
<span id="cb13-3"><a href="#cb13-3"></a>        <span class="dt">trimLeft =</span> <span class="dv">10</span>,  <span class="co"># this is the default</span></span>
<span id="cb13-4"><a href="#cb13-4"></a>        <span class="dt">trunLen =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">240</span>, <span class="dv">180</span>),  <span class="co"># forward and reverse truncation</span></span>
<span id="cb13-5"><a href="#cb13-5"></a>        <span class="dt">out_dir =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/tempfile">tempdir</a></span>()  <span class="co"># will store filtered files in a temporary dir</span></span>
<span id="cb13-6"><a href="#cb13-6"></a>    ),</span>
<span id="cb13-7"><a href="#cb13-7"></a>    <span class="dt">denoise =</span> <span class="kw"><a href="../reference/config_denoise.html">config_denoise</a></span>()  <span class="co"># will only use defaults</span></span>
<span id="cb13-8"><a href="#cb13-8"></a>)</span>
<span id="cb13-9"><a href="#cb13-9"></a></span>
<span id="cb13-10"><a href="#cb13-10"></a>denoised &lt;-<span class="st"> </span>quals <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/preprocess.html">preprocess</a></span>(config<span class="op">$</span>preprocess) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/denoise.html">denoise</a></span>(config<span class="op">$</span>denoise)</span></code></pre></div>
<pre><code>## INFO [2019-11-05 08:12:17] Preprocessing reads for 5 paired-end samples...
## INFO [2019-11-05 08:12:24] 3.23e+04/4.48e+04 (73.36%) reads passed preprocessing.
## INFO [2019-11-05 08:12:24] Running DADA2 on 1 run(s) from a sample of up to 2.5e+08 bases.
## INFO [2019-11-05 08:12:24] Learning errors for run `all` (5 samples)...
## INFO [2019-11-05 08:13:14] Dereplicating run `all` (5 samples)...
## INFO [2019-11-05 08:13:15] Inferring sequence variants for run `all`...
## INFO [2019-11-05 08:13:27] Finished run `all`.
## INFO [2019-11-05 08:13:27] Merged sequence tables, now removing chimeras...
## INFO [2019-11-05 08:13:27] Removed 5/164 sequence variants as chimeric (1.20% of reads)
## INFO [2019-11-05 08:13:27] Assigning taxonomy to 159 sequences...
## INFO [2019-11-05 08:13:27] Downloading taxa db to /var/folders/55/dv0p21y96g1cq84sr1zd3kym0000gr/T//RtmpRjeSft/taxa.fna.gz...
## INFO [2019-11-05 08:15:26] Downloading species db to /var/folders/55/dv0p21y96g1cq84sr1zd3kym0000gr/T//RtmpRjeSft/species.fna.gz...
## INFO [2019-11-05 08:16:19] Reads with taxonomic classification: Kingdom = 100 %,Phylum = 100 %,Class = 100 %,Order = 100 %,Family = 97 %,Genus = 56.9 %,Species = 8.3 %
## INFO [2019-11-05 08:16:19] Hashing 159 sequence variants.</code></pre>
<p>This will run both workflow steps sequentially and will use all available CPUs by default. You will get some disgnostic output on the logging interface but you can also inspect those in the returned artifact.</p>
<p>For instance to see how many reads were conserved in each processing step:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>denoised<span class="op">$</span>passed_reads</span></code></pre></div>
<pre><code>##      raw preprocessed   id derep_forward derep_reverse denoised_forward
## 1:  7793         5616 F3D0          5616          5616             5461
## 2:  5869         4388 F3D1          4388          4388             4280
## 3: 19620        13787 F3D2         13787         13787            13633
## 4:  6758         4527 F3D3          4527          4527             4424
## 5:  4779         3953 Mock          3953          3953             3933
##    denoised_reverse merged non_chimera
## 1:             5415   4795        4795
## 2:             4255   4005        4005
## 3:            13583  12872       12701
## 4:             4328   3761        3579
## 5:             3934   3895        3895</code></pre>
<p>We see that most samples were conserved (as reported in the logging).</p>
<p>The abundance of each sequence variant is saved in <code>denoised$feature_tables</code> which contains abundances for samples x ASVs. The ASVs here have been named by their md5 hash value and you can get the taxonomy and original sequence from <code>denoised$taxonomy</code>:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>denoised<span class="op">$</span>taxonomy[<span class="dv">1</span>, , drop =<span class="st"> </span><span class="ot">FALSE</span>]</span></code></pre></div>
<pre><code>##                                  Kingdom    Phylum          Class        
## c69a3db92e1dcdebb01dfaa5b5795559 "Bacteria" "Bacteroidetes" "Bacteroidia"
##                                  Order           Family           Genus
## c69a3db92e1dcdebb01dfaa5b5795559 "Bacteroidales" "Muribaculaceae" NA   
##                                  Species
## c69a3db92e1dcdebb01dfaa5b5795559 NA     
##                                  sequence                                                                                                                                                                                                                                                    
## c69a3db92e1dcdebb01dfaa5b5795559 "ACGGAGGATGCGAGCGTTATCCGGATTTATTGGGTTTAAAGGGTGCGCAGGCGGAAGATCAAGTCAGCGGTAAAATTGAGAGGCTCAACCTCTTCGAGCCGTTGAAACTGGTTTTCTTGAGTGAGCGAGAAGTATGCGGAATGCGTGGTGTAGCGGTGAAATGCATAGATATCACGCAGAACTCCGATTGCGAAGGCAGCATACCGGCGCTCAACTGACGCTCATGCACGAAAGTGTGGGTATCGAACAG"</code></pre>
<p>To see all steps run to obatin that output you can check the provenance:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>denoised<span class="op">$</span>steps</span></code></pre></div>
<pre><code>## [1] "quality_control" "preprocess"      "denoise"</code></pre>
</div>
<div id="managing-several-sequencing-runs" class="section level2">
<h2 class="hasAnchor">
<a href="#managing-several-sequencing-runs" class="anchor"></a>Managing several sequencing runs</h2>
<p>The <code>denoise</code> workflow step supports multiple sequencing runs in the same experiment. This requires that your files data table has a “run” column. If you sequencing files are organized such that every subfolder is a run you can simply use <code><a href="../reference/find_read_files.html">find_read_files(path, dirs_are_runs = TRUE)</a></code>. Otherwise, you have to specify the run by hand. We will do so for our files here:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>files &lt;-<span class="st"> </span>quals<span class="op">$</span>files</span>
<span id="cb21-2"><a href="#cb21-2"></a>files[, run <span class="op">:</span><span class="er">=</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="dt">times =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">1</span>))]</span>
<span id="cb21-3"><a href="#cb21-3"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(files)</span></code></pre></div>
<pre><code>##                                                                                                              forward
## 1: /Library/Frameworks/R.framework/Versions/3.6/Resources/library/mbtools/extdata/16S/F3D0_S188_L001_R1_001.fastq.gz
## 2: /Library/Frameworks/R.framework/Versions/3.6/Resources/library/mbtools/extdata/16S/F3D1_S189_L001_R1_001.fastq.gz
## 3: /Library/Frameworks/R.framework/Versions/3.6/Resources/library/mbtools/extdata/16S/F3D2_S190_L001_R1_001.fastq.gz
## 4: /Library/Frameworks/R.framework/Versions/3.6/Resources/library/mbtools/extdata/16S/F3D3_S191_L001_R1_001.fastq.gz
## 5: /Library/Frameworks/R.framework/Versions/3.6/Resources/library/mbtools/extdata/16S/Mock_S280_L001_R1_001.fastq.gz
##                                                                                                              reverse
## 1: /Library/Frameworks/R.framework/Versions/3.6/Resources/library/mbtools/extdata/16S/F3D0_S188_L001_R2_001.fastq.gz
## 2: /Library/Frameworks/R.framework/Versions/3.6/Resources/library/mbtools/extdata/16S/F3D1_S189_L001_R2_001.fastq.gz
## 3: /Library/Frameworks/R.framework/Versions/3.6/Resources/library/mbtools/extdata/16S/F3D2_S190_L001_R2_001.fastq.gz
## 4: /Library/Frameworks/R.framework/Versions/3.6/Resources/library/mbtools/extdata/16S/F3D3_S191_L001_R2_001.fastq.gz
## 5: /Library/Frameworks/R.framework/Versions/3.6/Resources/library/mbtools/extdata/16S/Mock_S280_L001_R2_001.fastq.gz
##      id injection_order lane run
## 1: F3D0             188    1   1
## 2: F3D1             189    1   1
## 3: F3D2             190    1   2
## 4: F3D3             191    1   2
## 5: Mock             280    1   3</code></pre>
<p>Rerunning the previous workflow with this file list will now give a slightly different output:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>byrun &lt;-<span class="st"> </span>files <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/preprocess.html">preprocess</a></span>(config<span class="op">$</span>preprocess) <span class="op">%&gt;%</span></span>
<span id="cb23-2"><a href="#cb23-2"></a><span class="st">         </span><span class="kw"><a href="../reference/denoise.html">denoise</a></span>(config<span class="op">$</span>denoise)</span></code></pre></div>
<pre><code>## INFO [2019-11-05 08:16:19] Preprocessing reads for 5 paired-end samples...
## INFO [2019-11-05 08:16:25] 3.23e+04/4.48e+04 (73.36%) reads passed preprocessing.
## INFO [2019-11-05 08:16:25] Running DADA2 on 3 run(s) from a sample of up to 2.5e+08 bases.
## INFO [2019-11-05 08:16:25] Learning errors for run `1` (2 samples)...
## INFO [2019-11-05 08:16:43] Dereplicating run `1` (2 samples)...
## INFO [2019-11-05 08:16:43] Inferring sequence variants for run `1`...
## INFO [2019-11-05 08:16:48] Finished run `1`.
## INFO [2019-11-05 08:16:48] Learning errors for run `2` (2 samples)...
## INFO [2019-11-05 08:17:13] Dereplicating run `2` (2 samples)...
## INFO [2019-11-05 08:17:14] Inferring sequence variants for run `2`...
## INFO [2019-11-05 08:17:20] Finished run `2`.
## INFO [2019-11-05 08:17:20] Learning errors for run `3` (1 samples)...
## INFO [2019-11-05 08:17:21] Dereplicating run `3` (1 samples)...
## INFO [2019-11-05 08:17:22] Inferring sequence variants for run `3`...
## INFO [2019-11-05 08:17:22] Finished run `3`.
## INFO [2019-11-05 08:17:22] Merged sequence tables, now removing chimeras...
## INFO [2019-11-05 08:17:22] Removed 5/162 sequence variants as chimeric (1.20% of reads)
## INFO [2019-11-05 08:17:22] Assigning taxonomy to 157 sequences...
## INFO [2019-11-05 08:17:22] Found existing taxa db `/var/folders/55/dv0p21y96g1cq84sr1zd3kym0000gr/T//RtmpRjeSft/taxa.fna.gz`. Using that one.
## INFO [2019-11-05 08:19:03] Found existing species db `/var/folders/55/dv0p21y96g1cq84sr1zd3kym0000gr/T//RtmpRjeSft/species.fna.gz`. Using that one.
## INFO [2019-11-05 08:19:33] Reads with taxonomic classification: Kingdom = 100 %,Phylum = 100 %,Class = 100 %,Order = 100 %,Family = 97 %,Genus = 57.6 %,Species = 8.2 %
## INFO [2019-11-05 08:19:33] Hashing 157 sequence variants.</code></pre>
<p>Error profiles will now be estimated for each run individually and final tables will be merged at the end. This is slower than having all files in one run so you should always try to distribute your samples across as few sequencing runs as possible.</p>
</div>
<div id="converting-to-phyloseq" class="section level2">
<h2 class="hasAnchor">
<a href="#converting-to-phyloseq" class="anchor"></a>Converting to phyloseq</h2>
<p>For all further analyses you might want to work with a data type which is more suited for amplicon abundance data. You can directly obtain a phyloseq object with</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a>ps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/as_phyloseq.html">as_phyloseq</a></span>(denoised)</span></code></pre></div>
<p>It is possible to pass in metadata as a data frame with <code>to_phyloseq(object, metadata = df)</code>.</p>
<p>You can now use all additional functions from <code>mbtools</code> on that. For instance we can get the ASV counts in a tidy data table:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>asv_counts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/taxa_count.html">taxa_count</a></span>(ps, <span class="dt">lev =</span> <span class="ot">NA</span>)</span>
<span id="cb26-2"><a href="#cb26-2"></a>asv_counts[sample <span class="op">==</span><span class="st"> "Mock"</span> <span class="op">&amp;</span><span class="st"> </span>reads <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>]</span></code></pre></div>
<pre><code>##     sample                             taxa reads
##  1:   Mock e941355a23a624aef3fe3e4caf9bc301   564
##  2:   Mock 532c8d14997a65c94cdabbf41c8f2647    90
##  3:   Mock 35b4a92774d4fe5a2d33367602b223a8   378
##  4:   Mock b2df6175ad0400ad1301072c5d64a75f   348
##  5:   Mock cc36ce261772980980931b51d3d9a379   309
##  6:   Mock 7b106324ebe9486ccd49abc287a6de45   288
##  7:   Mock ed8390cd975c833a487e2678b248bf13   272
##  8:   Mock 454886158d7f3bc1b770132e75a1c87f   243
##  9:   Mock ea98650694b6343f2ab88943711c2595   206
## 10:   Mock 5ff1477099ca90cc5c48b7b231b735eb   203
## 11:   Mock 75d2399d0d5f38dcd98e7f4c91271272   173
## 12:   Mock bf4c86674249ac9c02f5701d3dfdfd31   163
## 13:   Mock 8a214389774631998cca74c5b0088d8b   148
## 14:   Mock 37abc7707b7be944fb565da5297d3175   136
## 15:   Mock f43b31caaa0047cf889f30ec8e56ec17   131
## 16:   Mock 664acad52badcb22b536dd609d7c4f24   111
## 17:   Mock e0c48dcc0eb1e70ac90edb8107d4e0d4    69
## 18:   Mock b6a0a5f9e9187f987e31ebf1c8c4bd1d    48
## 19:   Mock b6f891b70ee28c80ed84667ea428baf3    15
##                            species
##  1:              Staphylococcus NA
##  2:               Lactobacillus NA
##  3:               Acinetobacter NA
##  4:                    Bacillus NA
##  5:            Helicobacter pylori
##  6: Clostridium_sensu_stricto_1 NA
##  7:         Neisseria meningitidis
##  8:      Actinomyces odontolyticus
##  9:           Bacteroides vulgatus
## 10:           Streptococcus mutans
## 11:                Enterococcus NA
## 12:        Escherichia/Shigella NA
## 13:                    Listeria NA
## 14:                 Pseudomonas NA
## 15:               Streptococcus NA
## 16:        Deinococcus radiodurans
## 17:       Porphyromonas gingivalis
## 18:                 Rhodobacter NA
## 19:               Streptococcus NA</code></pre>
<p>Here we see that we have identified 10 ASVs in the Mock sample. The exact number of strains in that sample is 20, which we don’t find since we don’t have enough reads available to learn a perfect error model. This is not surprising since those are only a few files from that particular study.</p>
<p>We could also plot the phyla distribution across the samples:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a><span class="kw"><a href="../reference/plot_taxa.html">plot_taxa</a></span>(ps)</span></code></pre></div>
<pre><code>## Warning in `[.data.table`(counts, , `:=`(id, as.numeric(sample))):
## Invalid .internal.selfref detected and fixed by taking a (shallow) copy
## of the data.table so that := can add this new column by reference. At
## an earlier point, this data.table has been copied by R (or was created
## manually using structure() or similar). Avoid names&lt;- and attr&lt;- which
## in R currently (and oddly) may copy the whole data.table. Use set* syntax
## instead to avoid copying: ?set, ?setnames and ?setattr. If this message
## doesn't help, please report your use case to the data.table issue tracker
## so the root cause can be fixed or this message improved.</code></pre>
<p><img src="02_amplicon_sequencing_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#amplicon-sequencing-data-analysis">Amplicon sequencing data analysis</a><ul class="nav nav-pills nav-stacked">
<li><a href="#quality-assessment">Quality assessment</a></li>
      <li><a href="#denoise">Denoise</a></li>
      <li><a href="#managing-several-sequencing-runs">Managing several sequencing runs</a></li>
      <li><a href="#converting-to-phyloseq">Converting to phyloseq</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Christian Diener.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
