<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>3 - Reference counting • miso</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="3 - Reference counting">
<meta property="og:description" content="miso">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">miso</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.51.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/00_qa.html">0 - Quality control of sequencing data</a>
    </li>
    <li>
      <a href="../articles/01_preprocessing.html">1 - Preprocessing of sequencing data</a>
    </li>
    <li>
      <a href="../articles/02_amplicon_sequencing.html">2 - Amplicon sequencing analysis</a>
    </li>
    <li>
      <a href="../articles/03_counting.html">3 - Reference counting</a>
    </li>
    <li>
      <a href="../articles/04_sra.html">4 - Submitting data to SRA</a>
    </li>
    <li>
      <a href="../articles/05_plate_layouts.html">5 - Plate Layouts</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dienerlab/miso/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>3 - Reference counting</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dienerlab/miso/blob/HEAD/vignettes/03_counting.Rmd" class="external-link"><code>vignettes/03_counting.Rmd</code></a></small>
      <div class="hidden name"><code>03_counting.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="reference-counting">Reference counting<a class="anchor" aria-label="anchor" href="#reference-counting"></a>
</h2>
<p>One basic anaylisis we want to perform is to count the abundance of
some references in our data. Here reference is a broad term and may
refer to transcripts, genomes, or amplicon sequences contained in a
reference database.</p>
<p>For this <code>miso</code> implements a general purpose expectation
maximization (EM) counter. It is implemented in C++ to perform fast and
usually reading the alignment file takes longer than the actual
analysis.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dienerlab/miso" class="external-link">miso</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Also loading:</span></span></code></pre>
<pre><code><span><span class="co">##   - dada2=1.28.0</span></span>
<span><span class="co">##   - data.table=1.15.2</span></span>
<span><span class="co">##   - ggplot2=3.5.0</span></span>
<span><span class="co">##   - magrittr=2.0.3</span></span>
<span><span class="co">##   - phyloseq=1.44.0</span></span>
<span><span class="co">##   - ShortRead=1.58.0</span></span>
<span><span class="co">##   - yaml=2.3.8</span></span></code></pre>
<pre><code><span><span class="co">## Found tools:</span></span></code></pre>
<pre><code><span><span class="co">##   - minimap2=2.27-r1193</span></span>
<span><span class="co">##   - samtools=1.19.2</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'miso'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked _by_ 'package:BiocGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     normalize</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:graphics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     layout</span></span></code></pre>
<div class="section level3">
<h3 id="alignment">Alignment<a class="anchor" aria-label="anchor" href="#alignment"></a>
</h3>
<p>Let’s start by aligning our short read data to a reference database
of 10 microbial genomes.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">alns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/aln"</span>, package <span class="op">=</span> <span class="st">"miso"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>      <span class="fu"><a href="../reference/find_alignments.html">find_alignments</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/genomes/zymo_mock.fna.gz"</span>,</span>
<span>                   package <span class="op">=</span> <span class="st">"miso"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="em-counting">EM counting<a class="anchor" aria-label="anchor" href="#em-counting"></a>
</h3>
<p>Counting is yet again a workflow and requires an alignment artifact
and a configuration.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">config</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/config_count.html">config_count</a></span><span class="op">(</span></span>
<span>    reference <span class="op">=</span> <span class="va">ref</span>,</span>
<span>    threads <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    weights <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">config</span></span></code></pre></div>
<pre><code><span><span class="co">## $reference</span></span>
<span><span class="co">## [1] "/tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/genomes/zymo_mock.fna.gz"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $threads</span></span>
<span><span class="co">## [1] 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $method</span></span>
<span><span class="co">## [1] "em"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $maxit</span></span>
<span><span class="co">## [1] 10000</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $cutoff</span></span>
<span><span class="co">## [1] 0.01</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $tpm</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $weights</span></span>
<span><span class="co">## [1] TRUE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attr(,"class")</span></span>
<span><span class="co">## [1] "config"</span></span></code></pre>
<p>And we can proceed to counting:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/count_references.html">count_references</a></span><span class="op">(</span><span class="va">alns</span>, <span class="va">config</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## INFO [2024-04-29 09:43:56] Getting reference lengths from /tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/genomes/zymo_mock.fna.gz...</span></span>
<span><span class="co">## INFO [2024-04-29 09:43:56] Normalized IDs. Starting counting...</span></span>
<span><span class="co">## INFO [2024-04-29 09:43:56] [/tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/aln/even1.bam] Read 16943 alignments.</span></span>
<span><span class="co">## INFO [2024-04-29 09:43:56] [/tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/aln/even1.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].</span></span>
<span><span class="co">## INFO [2024-04-29 09:43:56] [/tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/aln/even1.bam] Used 20 EM iterations on 16 equivalence classes. Last max. abs. change was 1.1e-13. Database concordance is 99.84%.</span></span>
<span><span class="co">## INFO [2024-04-29 09:43:56] [/tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/aln/even2.bam] Read 16803 alignments.</span></span>
<span><span class="co">## INFO [2024-04-29 09:43:56] [/tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/aln/even2.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].</span></span>
<span><span class="co">## INFO [2024-04-29 09:43:56] [/tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/aln/even2.bam] Used 20 EM iterations on 12 equivalence classes. Last max. abs. change was 2.3e-13. Database concordance is 99.83%.</span></span>
<span><span class="co">## INFO [2024-04-29 09:43:56] [/tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/aln/even3.bam] Read 17517 alignments.</span></span>
<span><span class="co">## INFO [2024-04-29 09:43:57] [/tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/aln/even3.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].</span></span>
<span><span class="co">## INFO [2024-04-29 09:43:57] [/tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/aln/even3.bam] Used 20 EM iterations on 20 equivalence classes. Last max. abs. change was 2.3e-13. Database concordance is 99.84%.</span></span></code></pre>
<p>This creates a count artifact which contains the counts.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cn</span><span class="op">$</span><span class="va">counts</span></span></code></pre></div>
<pre><code><span><span class="co">##      reference    counts effective_length sample</span></span>
<span><span class="co">##         &lt;char&gt;     &lt;num&gt;            &lt;num&gt; &lt;char&gt;</span></span>
<span><span class="co">##  1: AE004091.2 633.55872          6264306  even1</span></span>
<span><span class="co">##  2: AE006468.2 670.36751          4857352  even1</span></span>
<span><span class="co">##  3: AE016830.1 660.11661          3217933  even1</span></span>
<span><span class="co">##  4: AE017341.1 603.13777         19052864  even1</span></span>
<span><span class="co">##  5: AL009126.3 627.86614          4215508  even1</span></span>
<span><span class="co">##  6: AL591824.1 627.88684          2944430  even1</span></span>
<span><span class="co">##  7: AP008937.1 463.49491          2098587  even1</span></span>
<span><span class="co">##  8: BA000007.3 701.72676          5498480  even1</span></span>
<span><span class="co">##  9: BK006935.2 583.40000         12072428  even1</span></span>
<span><span class="co">## 10: CP000253.1 717.15854          2821263  even1</span></span>
<span><span class="co">## 11:       &lt;NA&gt;  10.28620               NA  even1</span></span>
<span><span class="co">## 12: AE004091.2 637.96413          6264306  even2</span></span>
<span><span class="co">## 13: AE006468.2 546.16769          4857352  even2</span></span>
<span><span class="co">## 14: AE016830.1 684.69335          3217933  even2</span></span>
<span><span class="co">## 15: AE017341.1 588.45604         19052864  even2</span></span>
<span><span class="co">## 16: AL009126.3 587.77636          4215508  even2</span></span>
<span><span class="co">## 17: AL591824.1 510.05621          2944430  even2</span></span>
<span><span class="co">## 18: AP008937.1 670.60797          2098587  even2</span></span>
<span><span class="co">## 19: BA000007.3 630.95856          5498480  even2</span></span>
<span><span class="co">## 20: BK006935.2 608.11643         12072428  even2</span></span>
<span><span class="co">## 21: CP000253.1 590.10050          2821263  even2</span></span>
<span><span class="co">## 22:       &lt;NA&gt;  10.10276               NA  even2</span></span>
<span><span class="co">## 23: AE004091.2 616.09439          6264306  even3</span></span>
<span><span class="co">## 24: AE006468.2 653.06348          4857352  even3</span></span>
<span><span class="co">## 25: AE016830.1 600.05374          3217933  even3</span></span>
<span><span class="co">## 26: AE017341.1 638.01088         19052864  even3</span></span>
<span><span class="co">## 27: AL009126.3 556.39173          4215508  even3</span></span>
<span><span class="co">## 28: AL591824.1 718.93137          2944430  even3</span></span>
<span><span class="co">## 29: AP008937.1 731.04016          2098587  even3</span></span>
<span><span class="co">## 30: BA000007.3 553.07477          5498480  even3</span></span>
<span><span class="co">## 31: BK006935.2 610.69845         12072428  even3</span></span>
<span><span class="co">## 32: CP000253.1 485.63165          2821263  even3</span></span>
<span><span class="co">## 33:       &lt;NA&gt;  10.00938               NA  even3</span></span>
<span><span class="co">##      reference    counts effective_length sample</span></span></code></pre>
<p>Note that there is a NA reference. This is the number of reads likely
coming from a sequence not contained in the reference.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cn</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">reference</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##    reference   counts effective_length sample</span></span>
<span><span class="co">##       &lt;char&gt;    &lt;num&gt;            &lt;num&gt; &lt;char&gt;</span></span>
<span><span class="co">## 1:      &lt;NA&gt; 10.28620               NA  even1</span></span>
<span><span class="co">## 2:      &lt;NA&gt; 10.10276               NA  even2</span></span>
<span><span class="co">## 3:      &lt;NA&gt; 10.00938               NA  even3</span></span></code></pre>
<p>In each sample we have 10 microbes with the same abundance.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="va">cn</span><span class="op">$</span><span class="va">counts</span>, <span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">counts</span>, y<span class="op">=</span><span class="va">reference</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">xlim</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1000</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">sample</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_counting_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>Looks ok, but there is a lot of noise since we have very low
depth.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="va">cn</span><span class="op">$</span><span class="va">counts</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">counts</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_histogram</span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span></code></pre></div>
<p><img src="03_counting_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>We could also compare that with a naive counting method that just
uses the best alignment score. In our case that will perform equally
since we have little multi-mapping here.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">config</span><span class="op">$</span><span class="va">method</span> <span class="op">&lt;-</span> <span class="st">"naive"</span></span>
<span><span class="va">cn2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/count_references.html">count_references</a></span><span class="op">(</span><span class="va">alns</span>, <span class="va">config</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## INFO [2024-04-29 09:43:57] Getting reference lengths from /tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/genomes/zymo_mock.fna.gz...</span></span>
<span><span class="co">## INFO [2024-04-29 09:43:57] Normalized IDs. Starting counting...</span></span>
<span><span class="co">## INFO [2024-04-29 09:43:57] [/tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/aln/even1.bam] Read 16943 alignments.</span></span>
<span><span class="co">## INFO [2024-04-29 09:43:57] [/tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/aln/even1.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].</span></span>
<span><span class="co">## INFO [2024-04-29 09:43:57] [/tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/aln/even2.bam] Read 16803 alignments.</span></span>
<span><span class="co">## INFO [2024-04-29 09:43:57] [/tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/aln/even2.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].</span></span>
<span><span class="co">## INFO [2024-04-29 09:43:57] [/tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/aln/even3.bam] Read 17517 alignments.</span></span>
<span><span class="co">## INFO [2024-04-29 09:43:57] [/tmp/RtmpFnpEcH/temp_libpath39b6c5ce98287/miso/extdata/aln/even3.bam] 10 reference seqs. Confidence interval for effective lengths: [2261189.10, 17482265.90].</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">cn</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="va">cn2</span><span class="op">$</span><span class="va">counts</span>, on <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample"</span>, <span class="st">"reference"</span>, <span class="st">"effective_length"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">counts</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">counts</span>, y <span class="op">=</span> <span class="va">i.counts</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_smooth</span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"em"</span>, y <span class="op">=</span> <span class="st">"naive"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme_classic</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `geom_smooth()` using formula = 'y ~ x'</span></span></code></pre>
<p><img src="03_counting_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>As we see they both give the same results for that simple case.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Christian Diener.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
