<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>4 - Submitting data to SRA â€¢ miso</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="4 - Submitting data to SRA">
<meta property="og:description" content="miso">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">miso</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.50.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/00_qa.html">0 - Quality control of sequencing data</a>
    </li>
    <li>
      <a href="../articles/01_preprocessing.html">1 - Preprocessing of sequencing data</a>
    </li>
    <li>
      <a href="../articles/02_amplicon_sequencing.html">2 - Amplicon sequencing analysis</a>
    </li>
    <li>
      <a href="../articles/03_counting.html">3 - Reference counting</a>
    </li>
    <li>
      <a href="../articles/04_sra.html">4 - Submitting data to SRA</a>
    </li>
    <li>
      <a href="../articles/05_plate_layouts.html">5 - Plate Layouts</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dienerlab/miso/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>4 - Submitting data to SRA</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dienerlab/miso/blob/HEAD/vignettes/04_sra.Rmd" class="external-link"><code>vignettes/04_sra.Rmd</code></a></small>
      <div class="hidden name"><code>04_sra.Rmd</code></div>

    </div>

    
    
<p>At some point of your project youy will want to make your data
available to the public. This often required to happen in the form of a
submission to the Sequence Read Archive (SRA), which can be challenging.
<code>miso</code> includes helpers to prep a submission for you.</p>
<p>In order to create a submission you will need the following:</p>
<ol style="list-style-type: decimal">
<li>A file manifest for all of your sequencing data.</li>
<li>A metadata table describing the sequenced samples.</li>
<li>Some basic information about the study and sequencing.</li>
</ol>
<p>Number 1 is managed as in all other <code>miso</code> steps in that
it is a file list such as used in quality control etc. For instance for
our example data:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dienerlab/miso" class="external-link">miso</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Also loading:</span></span></code></pre>
<pre><code><span><span class="co">##   - dada2=1.28.0</span></span>
<span><span class="co">##   - data.table=1.15.2</span></span>
<span><span class="co">##   - ggplot2=3.5.0</span></span>
<span><span class="co">##   - magrittr=2.0.3</span></span>
<span><span class="co">##   - phyloseq=1.44.0</span></span>
<span><span class="co">##   - ShortRead=1.58.0</span></span>
<span><span class="co">##   - yaml=2.3.8</span></span></code></pre>
<pre><code><span><span class="co">## Found tools:</span></span></code></pre>
<pre><code><span><span class="co">##   - minimap2=2.27-r1193</span></span>
<span><span class="co">##   - samtools=1.19.2</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'miso'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked _by_ 'package:BiocGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     normalize</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:graphics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     layout</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/shotgun"</span>, package <span class="op">=</span> <span class="st">"miso"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/find_read_files.html">find_read_files</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Which gives you:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fi</span></span></code></pre></div>
<pre><code><span><span class="co">## Key: &lt;id&gt;</span></span>
<span><span class="co">##                                                                                          forward</span></span>
<span><span class="co">##                                                                                           &lt;char&gt;</span></span>
<span><span class="co">## 1: /tmp/Rtmp0LsDXk/temp_libpath20758c3f036fbb/miso/extdata/shotgun/even1_S1_L001_R1_001.fasta.gz</span></span>
<span><span class="co">## 2: /tmp/Rtmp0LsDXk/temp_libpath20758c3f036fbb/miso/extdata/shotgun/even2_S2_L001_R1_001.fasta.gz</span></span>
<span><span class="co">## 3: /tmp/Rtmp0LsDXk/temp_libpath20758c3f036fbb/miso/extdata/shotgun/even3_S3_L001_R1_001.fasta.gz</span></span>
<span><span class="co">##                                                                                          reverse</span></span>
<span><span class="co">##                                                                                           &lt;char&gt;</span></span>
<span><span class="co">## 1: /tmp/Rtmp0LsDXk/temp_libpath20758c3f036fbb/miso/extdata/shotgun/even1_S1_L001_R2_001.fasta.gz</span></span>
<span><span class="co">## 2: /tmp/Rtmp0LsDXk/temp_libpath20758c3f036fbb/miso/extdata/shotgun/even2_S2_L001_R2_001.fasta.gz</span></span>
<span><span class="co">## 3: /tmp/Rtmp0LsDXk/temp_libpath20758c3f036fbb/miso/extdata/shotgun/even3_S3_L001_R2_001.fasta.gz</span></span>
<span><span class="co">##        id injection_order  lane</span></span>
<span><span class="co">##    &lt;char&gt;           &lt;num&gt; &lt;num&gt;</span></span>
<span><span class="co">## 1:  even1               1     1</span></span>
<span><span class="co">## 2:  even2               2     1</span></span>
<span><span class="co">## 3:  even3               3     1</span></span></code></pre>
<p>The metadata is a data frame describing additional attributes for
each of those samples. For instance you could use the
<code>sample_data()</code> slot of a phyloseq object. However, this
table always needs to have a column specifying the sample ID as in the
file manifest and a date column denoting the date of sample extraction.
Letâ€™s create one for the three samples in our example data:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="op">(</span></span>
<span>    id <span class="op">=</span> <span class="va">fi</span><span class="op">$</span><span class="va">id</span>,</span>
<span>    date <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"2019-01-01"</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>    diet <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"vegan"</span>, <span class="st">"vegetarian"</span>, <span class="st">"flexitarian"</span><span class="op">)</span>,</span>
<span>    age <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">23</span>, <span class="fl">38</span>, <span class="fl">64</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>For the additional study data <code>miso</code> will use a set of
presets that will set a lot of the information for you but you will
still need to specify additional aspects of your study. This is again
managed by configurations and you can see the set of configuration
variables by creating one with <code>config_sra</code>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/config_sra.html">config_sra</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## $metadata</span></span>
<span><span class="co">## NULL</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $id_col</span></span>
<span><span class="co">## [1] "id"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $date_col</span></span>
<span><span class="co">## [1] "date"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $country</span></span>
<span><span class="co">## [1] "USA"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $preset</span></span>
<span><span class="co">## [1] "human gut 16S"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $out_dir</span></span>
<span><span class="co">## [1] "sra"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $title</span></span>
<span><span class="co">## NULL</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $platform</span></span>
<span><span class="co">## [1] "ILLUMINA"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $instrument_model</span></span>
<span><span class="co">## [1] "Illumina MiSeq"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $bioproject</span></span>
<span><span class="co">## NULL</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $latitude</span></span>
<span><span class="co">## [1] 42.36</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $longitude</span></span>
<span><span class="co">## [1] -71.0941</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $make_package</span></span>
<span><span class="co">## [1] TRUE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attr(,"class")</span></span>
<span><span class="co">## [1] "config"</span></span></code></pre>
<p>Except for the <code>bioproject</code> (fill only if you already have
a BIOPROJECT ID) variable everything else needs to be specified (or the
default values will be used). You can check the the list of available
presets with:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">miso</span><span class="fu">:::</span><span class="va">presets</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "human gut 16S"         "mouse gut 16S"         "human gut metagenome" </span></span>
<span><span class="co">## [4] "human skin metagenome" "mouse gut metagenome"  "human gut RNA-Seq"    </span></span>
<span><span class="co">## [7] "mouse gut RNA-Seq"     "aquatic biofilm 16S"   "in vitro metagenome"</span></span></code></pre>
<p>Letâ€™s fill in the data for the example study:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">config</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/config_sra.html">config_sra</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Sequencing of the gut metagenome: the effect of age on Bacteroides"</span>,</span>
<span>    preset <span class="op">=</span> <span class="st">"human gut metagenome"</span>,</span>
<span>    platform <span class="op">=</span> <span class="st">"ILLUMINA"</span>,</span>
<span>    instrument <span class="op">=</span> <span class="st">"Illumina HiSeq 2000"</span>,</span>
<span>    metadata <span class="op">=</span> <span class="va">metadata</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>You can now proceed building the subsmission data:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sra</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sra_submission.html">sra_submission</a></span><span class="op">(</span><span class="va">fi</span>, <span class="va">config</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## INFO [2024-04-25 13:00:31] Packing submission files to sra/sra_files.tar.gz.</span></span>
<span><span class="co">## INFO [2024-04-25 13:00:31] Writing biosample attributes to sra/05_biosample_attributes.tsv.</span></span>
<span><span class="co">## INFO [2024-04-25 13:00:31] Writing file metadata to sra/06_sra_metadata.tsv.</span></span>
<span><span class="co">## INFO [2024-04-25 13:00:31] You are now ready for submission. Go to https://submit.ncbi.nlm.nih.gov/subs/sra/, log in and click on `New submission`. Fill in the general data for your project in steps 1 through 3. In step 4 choose `Environmental/Metagenome Genomic Sequences MIMS` with the appropriate environment. In step 5 and 6 upload the respective files in sra. In step 7 you can directly upload the `*.tar.gz` submission package. Just click on `continue` another time to have the archive unpacked as indicated.</span></span></code></pre>
<p>This told us exactly how to submit the data. We urge you to first
check all created files for correctness. This can be done by inspecting
the return value. For instance:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sra</span><span class="op">$</span><span class="va">sra_metadata</span></span></code></pre></div>
<p>When looking in the specified folder (<code>./sra</code> by default)
you will see the tables for step 5 and 6:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"sra"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "05_biosample_attributes.tsv" "06_sra_metadata.tsv"        </span></span>
<span><span class="co">## [3] "sra_files.tar.gz"</span></span></code></pre>
<p>For larger submission we recommend installing the Aspera client in
your browser as this will speed up the upload significantly and allow
for larger packages.</p>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Christian Diener.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
