<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>8 - Submitting data to SRA • mbtools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="8 - Submitting data to SRA">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mbtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.40.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/00_qa.html">0 - Quality control of sequencing data</a>
    </li>
    <li>
      <a href="../articles/01_preprocessing.html">1 - Preprocessing of sequencing data</a>
    </li>
    <li>
      <a href="../articles/02_amplicon_sequencing.html">2 - Amplicon sequencing analysis</a>
    </li>
    <li>
      <a href="../articles/03_shotgun_align.html">3 - Shotgun data - alignments</a>
    </li>
    <li>
      <a href="../articles/04_long_align.html">4 - Long read data - alignments</a>
    </li>
    <li>
      <a href="../articles/05_lineage.html">5 - Lineage calling with SLIMM</a>
    </li>
    <li>
      <a href="../articles/06_counting.html">6 - Reference counting</a>
    </li>
    <li>
      <a href="../articles/07_workflow_management.html">7 - Workflow management</a>
    </li>
    <li>
      <a href="../articles/08_sra.html">8 - Submitting data to SRA</a>
    </li>
    <li>
      <a href="../articles/09_plate_layouts.html">9 - Plate Layouts</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Gibbons-Lab/mbtools">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>8 - Submitting data to SRA</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Gibbons-Lab/mbtools/blob/master/vignettes/08_sra.Rmd"><code>vignettes/08_sra.Rmd</code></a></small>
      <div class="hidden name"><code>08_sra.Rmd</code></div>

    </div>

    
    
<p>At some point of your project youy will want to make your data available to the public. This often required to happen in the form of a submission to the Sequence Read Archive (SRA), which can be challenging. <code>mbtools</code> includes helpers to prep a submission for you.</p>
<p>In order to create a submission you will need the following:</p>
<ol style="list-style-type: decimal">
<li>A file manifest for all of your sequencing data.</li>
<li>A metadata table describing the sequenced samples.</li>
<li>Some basic information about the study and sequencing.</li>
</ol>
<p>Number 1 is managed as in all other <code>mbtools</code> steps in that it is a file list such as used in quality control etc. For instance for our example data:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(mbtools)</span></code></pre></div>
<pre><code>## Also loading:</code></pre>
<pre><code>##   - dada2=1.12.1
##   - data.table=1.12.6
##   - ggplot2=3.2.1
##   - magrittr=1.5
##   - phyloseq=1.28.0
##   - ShortRead=1.42.0
##   - yaml=2.2.0</code></pre>
<pre><code>## Found tools:</code></pre>
<pre><code>##   - minimap2=2.17-r941
##   - slimm=0.3.4
##   - samtools=1.9</code></pre>
<pre><code>## 
## Attaching package: 'mbtools'</code></pre>
<pre><code>## The following object is masked _by_ 'package:BiocGenerics':
## 
##     normalize</code></pre>
<pre><code>## The following object is masked from 'package:graphics':
## 
##     layout</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>fi &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata/shotgun"</span>, <span class="dt">package =</span> <span class="st">"mbtools"</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/find_read_files.html">find_read_files</a></span>()</span></code></pre></div>
<p>Which gives you:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>fi</span></code></pre></div>
<p>The metadata is a data frame describing additional attributes for each of those samples. For instance you could use the <code>sample_data()</code> slot of a phyloseq object. However, this table always needs to have a column specifying the sample ID as in the file manifest and a date column denoting the date of sample extraction. Let’s create one for the three samples in our example data:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>metadata &lt;-<span class="st"> </span><span class="kw">data.table</span>(</span>
<span id="cb11-2"><a href="#cb11-2"></a>    <span class="dt">id =</span> fi<span class="op">$</span>id,</span>
<span id="cb11-3"><a href="#cb11-3"></a>    <span class="dt">date =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="st">"2019-01-01"</span>, <span class="dv">3</span>),</span>
<span id="cb11-4"><a href="#cb11-4"></a>    <span class="dt">diet =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"vegan"</span>, <span class="st">"vegitarian"</span>, <span class="st">"flexitarian"</span>),</span>
<span id="cb11-5"><a href="#cb11-5"></a>    <span class="dt">age =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">23</span>, <span class="dv">38</span>, <span class="dv">64</span>)</span>
<span id="cb11-6"><a href="#cb11-6"></a>)</span></code></pre></div>
<p>For the additional study data <code>mbtools</code> will use a set of presets that will set a lot of the information for you but you will still need to specify additional aspects of your study. This is again managed by configurations and you can see the set of configuration variables by creating one with <code>config_sra</code>:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="kw"><a href="../reference/config_sra.html">config_sra</a></span>()</span></code></pre></div>
<pre><code>## $metadata
## NULL
## 
## $id_col
## [1] "id"
## 
## $date_col
## [1] "date"
## 
## $country
## [1] "USA"
## 
## $preset
## [1] "human gut 16S"
## 
## $out_dir
## [1] "sra"
## 
## $title
## NULL
## 
## $platform
## [1] "ILLUMINA"
## 
## $instrument_model
## [1] "Illumina MiSeq"
## 
## $bioproject
## NULL
## 
## $latitude
## [1] 42.36
## 
## $longitude
## [1] -71.0941
## 
## $make_package
## [1] TRUE
## 
## attr(,"class")
## [1] "config"</code></pre>
<p>Except for the <code>bioproject</code> (fill only if you already have a BIOPROJECT ID) variable everything else needs to be specified (or the default values will be used). You can check the the list of available presets with:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(mbtools<span class="op">:::</span>presets)</span></code></pre></div>
<pre><code>## [1] "human gut 16S"        "mouse gut 16S"        "human gut metagenome"
## [4] "mouse gut metagenome" "human gut RNA-Seq"    "mouse gut RNA-Seq"</code></pre>
<p>Let’s fill in the data for the example study:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>config &lt;-<span class="st"> </span><span class="kw"><a href="../reference/config_sra.html">config_sra</a></span>(</span>
<span id="cb16-2"><a href="#cb16-2"></a>    <span class="dt">title =</span> <span class="st">"Sequencing of the gut metagenome: the effect of age on Bacteroides"</span>,</span>
<span id="cb16-3"><a href="#cb16-3"></a>    <span class="dt">preset =</span> <span class="st">"human gut metagenome"</span>,</span>
<span id="cb16-4"><a href="#cb16-4"></a>    <span class="dt">platform =</span> <span class="st">"ILLUMINA"</span>,</span>
<span id="cb16-5"><a href="#cb16-5"></a>    <span class="dt">instrument =</span> <span class="st">"Illumina HiSeq 2000"</span>,</span>
<span id="cb16-6"><a href="#cb16-6"></a>    <span class="dt">metadata =</span> metadata</span>
<span id="cb16-7"><a href="#cb16-7"></a>)</span></code></pre></div>
<p>You can now proceed building the subsmission data:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a>sra &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sra_submission.html">sra_submission</a></span>(fi, config)</span></code></pre></div>
<pre><code>## INFO [2019-11-05 08:22:19] Packing submission files to sra/sra_files.tar.gz.
## INFO [2019-11-05 08:22:20] Writing biosample attributes to sra/05_biosample_attributes.tsv.
## INFO [2019-11-05 08:22:20] Writing file metadata to sra/06_sra_metadata.tsv.
## INFO [2019-11-05 08:22:20] You are now ready for submission. **Please make sure you only upload sequence files where human sequences have been removed and (you can use the `filter_reference` workflow for that)!**Go to https://submit.ncbi.nlm.nih.gov/subs/sra/, log in and click on `New submission`. Fill in the general data for your project in steps 1 through 3. In step 4 choose `Genome, metagenome or marker sequences (MIxS compliant)` and `Environmental/Metagenome Genomic Sequences MIMS`. In step 5 and 6 upload the respective files in sra. In step 7 you can directly upload the `*.tar.gz` submission package. Just click on `continue` another time to have the archive unpacked as indicated.</code></pre>
<p>This told us exactly how to submit the data. We urge you to first check all created files for correctness. This can be done by inspecting the return value. For insatnce:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>sra<span class="op">$</span>sra_metadata</span></code></pre></div>
<p>When looking in the specified folder (<code>./sra</code> by default) you will see the tables for step 5 and 6:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="st">"sra"</span>)</span></code></pre></div>
<pre><code>## [1] "05_biosample_attributes.tsv" "06_sra_metadata.tsv"        
## [3] "sra_files.tar.gz"</code></pre>
<p>For larger submission we recommend installing the Aspera client in your browser as this will speed up the upload significantly and allow for larger packages.</p>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Christian Diener.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
