<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ miso</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">miso</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.50.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/00_qa.html">0 - Quality control of sequencing data</a>
    </li>
    <li>
      <a href="../articles/01_preprocessing.html">1 - Preprocessing of sequencing data</a>
    </li>
    <li>
      <a href="../articles/02_amplicon_sequencing.html">2 - Amplicon sequencing analysis</a>
    </li>
    <li>
      <a href="../articles/03_counting.html">3 - Reference counting</a>
    </li>
    <li>
      <a href="../articles/04_sra.html">4 - Submitting data to SRA</a>
    </li>
    <li>
      <a href="../articles/05_plate_layouts.html">5 - Plate Layouts</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/dienerlab/miso/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="ERR260132.html">ERR260132</a></code> </p>
        </td>
        <td><p>Precomputed lineage calling and coverage example.</p></td>
      </tr><tr><td>
          <p><code><a href="annotate_contigs.html">annotate_contigs()</a></code> </p>
        </td>
        <td><p>Annotate diamond database hits with transcript info.</p></td>
      </tr><tr><td>
          <p><code><a href="annotate_silva.html">annotate_silva()</a></code> </p>
        </td>
        <td><p>Annotate the sequences in the SILVA database file.</p></td>
      </tr><tr><td>
          <p><code><a href="as.matrix.mbquant.html">as.matrix(<i>&lt;mbquant&gt;</i>)</a></code> </p>
        </td>
        <td><p>Convert a mbquant data table to a matrix.</p></td>
      </tr><tr><td>
          <p><code><a href="as_phyloseq.html">as_phyloseq()</a></code> </p>
        </td>
        <td><p>Converts the denoise artifact to a phyloseq object.</p></td>
      </tr><tr><td>
          <p><code><a href="association.html">association()</a></code> </p>
        </td>
        <td><p>Run differential association tests between taxa counts and exogenous factors.</p></td>
      </tr><tr><td>
          <p><code><a href="bracken_to_phyloseq.html">bracken_to_phyloseq()</a></code> </p>
        </td>
        <td><p>Converts BRACKEN results to a phyloseq object.</p></td>
      </tr><tr><td>
          <p><code><a href="build_lineage.html">build_lineage()</a></code> </p>
        </td>
        <td><p>Build the lineage from a SILVA path string.</p></td>
      </tr><tr><td>
          <p><code><a href="combinatorial_association.html">combinatorial_association()</a></code> </p>
        </td>
        <td><p>Run differential association tests between between all combinations of a factor variable with DESeq2. Can be used as post-hoc test for regression.</p></td>
      </tr><tr><td>
          <p><code><a href="config_association.html">config_association()</a></code> </p>
        </td>
        <td><p>Build a configuration for the alignment workflows.</p></td>
      </tr><tr><td>
          <p><code><a href="config_chimera.html">config_chimera()</a></code> </p>
        </td>
        <td><p>Build a configuration for chimera removal.</p></td>
      </tr><tr><td>
          <p><code><a href="config_count.html">config_count()</a></code> </p>
        </td>
        <td><p>Build a configuration for the transcript counting workflow.</p></td>
      </tr><tr><td>
          <p><code><a href="config_demultiplex.html">config_demultiplex()</a></code> </p>
        </td>
        <td><p>Build a configuration for the demultiplexing workflow.</p></td>
      </tr><tr><td>
          <p><code><a href="config_denoise.html">config_denoise()</a></code> </p>
        </td>
        <td><p>Build a configuration for the DADA2 workflow.</p></td>
      </tr><tr><td>
          <p><code><a href="config_layout.html">config_layout()</a></code> </p>
        </td>
        <td><p>Build a configuration for the plate layout.</p></td>
      </tr><tr><td>
          <p><code><a href="config_power.html">config_power()</a></code> </p>
        </td>
        <td><p>Build a configuration for power analysis.</p></td>
      </tr><tr><td>
          <p><code><a href="config_preprocess.html">config_preprocess()</a></code> </p>
        </td>
        <td><p>Build a configuration for raw read preprocessing.</p></td>
      </tr><tr><td>
          <p><code><a href="config_reference.html">config_reference()</a></code> </p>
        </td>
        <td><p>Build a configuration for the reference removal workflow.</p></td>
      </tr><tr><td>
          <p><code><a href="config_sra.html">config_sra()</a></code> </p>
        </td>
        <td><p>Build a configuration for the SRA submission workflow.</p></td>
      </tr><tr><td>
          <p><code><a href="count_references.html">count_references()</a></code> </p>
        </td>
        <td><p>Count alignment hits to a reference database.</p></td>
      </tr><tr><td>
          <p><code><a href="demultiplex.html">demultiplex()</a></code> </p>
        </td>
        <td><p>Splits FASTQ files into individual samples.</p></td>
      </tr><tr><td>
          <p><code><a href="denoise.html">denoise()</a></code> </p>
        </td>
        <td><p>Runs a full DADA2 workflow.</p></td>
      </tr><tr><td>
          <p><code><a href="discretize.html">discretize()</a></code> </p>
        </td>
        <td><p>Discretize all continuous variables in a data frame.</p></td>
      </tr><tr><td>
          <p><code><a href="download_bacterial_transcripts.html">download_bacterial_transcripts()</a></code> </p>
        </td>
        <td><p>Download the entire ENSEMBL transcript DB for bacteria.</p></td>
      </tr><tr><td>
          <p><code><a href="download_files.html">download_files()</a></code> </p>
        </td>
        <td><p>Download a list of files.</p></td>
      </tr><tr><td>
          <p><code><a href="download_sra.html">download_sra()</a></code> </p>
        </td>
        <td><p>Download data from SRA.</p></td>
      </tr><tr><td>
          <p><code><a href="effective_lengths.html">effective_lengths()</a></code> </p>
        </td>
        <td><p>Calculate the effective transcript lengths. This is the mean number of positions in the transcript the fragment could map to.</p></td>
      </tr><tr><td>
          <p><code><a href="em_count.html">em_count()</a></code> </p>
        </td>
        <td><p>Count transcripts using an Expectation Maximization (EM) algorithm.</p></td>
      </tr><tr><td>
          <p><code><a href="file_formats.html">file_formats</a></code> </p>
        </td>
        <td><p>Presets for read file formats.</p></td>
      </tr><tr><td>
          <p><code><a href="filter_counts.html">filter_counts()</a></code> </p>
        </td>
        <td><p>Filter low presence taxa from a count matrix.</p></td>
      </tr><tr><td>
          <p><code><a href="filter_reference.html">filter_reference()</a></code> </p>
        </td>
        <td><p>Filter a set of reference sequences from the data set.seed</p></td>
      </tr><tr><td>
          <p><code><a href="find_alignments.html">find_alignments()</a></code> </p>
        </td>
        <td><p>Find alignment BAM files in a given directory.</p></td>
      </tr><tr><td>
          <p><code><a href="find_read_files.html">find_read_files()</a></code> </p>
        </td>
        <td><p>Find read files in a given directory.</p></td>
      </tr><tr><td>
          <p><code><a href="find_taxa.html">find_taxa()</a></code> </p>
        </td>
        <td><p>Checks whether taxa from one taxonomy table are contained in another table.</p></td>
      </tr><tr><td>
          <p><code><a href="layout.html">layout()</a></code> </p>
        </td>
        <td><p>Generate a plate layout from a sample manifest.</p></td>
      </tr><tr><td>
          <p><code><a href="merge_transcripts.html">merge_transcripts()</a></code> </p>
        </td>
        <td><p>Merge transcripts into a single database</p></td>
      </tr><tr><td>
          <p><code><a href="miso.html">miso-package</a></code> <code><a href="miso.html">miso</a></code> </p>
        </td>
        <td><p>miso.</p></td>
      </tr><tr><td>
          <p><code><a href="mock_plot.html">mock_plot()</a></code> </p>
        </td>
        <td><p>Creates a plot of measured taxa quantifications vs. reference quantification.</p></td>
      </tr><tr><td>
          <p><code><a href="mockrobiota.html">mockrobiota()</a></code> </p>
        </td>
        <td><p>Downloads a complete data set from mockrobiota.</p></td>
      </tr><tr><td>
          <p><code><a href="mothur_to_dada.html">mothur_to_dada()</a></code> </p>
        </td>
        <td><p>Converts taxa annotations from mothur format to dada2 format.</p></td>
      </tr><tr><td>
          <p><code><a href="normalize.html">normalize()</a></code> </p>
        </td>
        <td><p>Normalize a set of read counts across samples</p></td>
      </tr><tr><td>
          <p><code><a href="orlitsky.html">orlitsky()</a></code> </p>
        </td>
        <td><p>Orlitsky's diminishing attenuation estimator (q2/3).</p></td>
      </tr><tr><td>
          <p><code><a href="parse_ensembl_id.html">parse_ensembl_id()</a></code> </p>
        </td>
        <td><p>Parse annotations from an ENSEMBL id</p></td>
      </tr><tr><td>
          <p><code><a href="plot_bb_fits.html">plot_bb_fits()</a></code> </p>
        </td>
        <td><p>Plot the Betabinomial fits for all taxa in a phyloseq object</p></td>
      </tr><tr><td>
          <p><code><a href="plot_counts.html">plot_counts()</a></code> </p>
        </td>
        <td><p>Plots counts for several taxa across a co-variable</p></td>
      </tr><tr><td>
          <p><code><a href="plot_entropy.html">plot_entropy()</a></code> </p>
        </td>
        <td><p>Plots the distribution of base entropy for each cycle.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_lengths.html">plot_lengths()</a></code> </p>
        </td>
        <td><p>Plots the distribution of sequence lengths with acceptable quality.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_qualities.html">plot_qualities()</a></code> </p>
        </td>
        <td><p>Plots the quality profile for an entire experiment.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_taxa.html">plot_taxa()</a></code> </p>
        </td>
        <td><p>Plots relative distribution for taxa across samples.</p></td>
      </tr><tr><td>
          <p><code><a href="power_analysis.html">power_analysis()</a></code> </p>
        </td>
        <td><p>Run a power analysis for microbe abundances from a reference sample.</p></td>
      </tr><tr><td>
          <p><code><a href="preprocess.html">preprocess()</a></code> </p>
        </td>
        <td><p>Runs preprocessing of sequencing reads.</p></td>
      </tr><tr><td>
          <p><code><a href="quality_control.html">quality_control()</a></code> </p>
        </td>
        <td><p>Workflow for basic quality assessment of the read files.</p></td>
      </tr><tr><td>
          <p><code><a href="quality_profile.html">quality_profile()</a></code> </p>
        </td>
        <td><p>Get the quality profiles and base calls for each cycle across all input files.</p></td>
      </tr><tr><td>
          <p><code><a href="read_bam.html">read_bam()</a></code> </p>
        </td>
        <td><p>Read alignments from a BAM file.</p></td>
      </tr><tr><td>
          <p><code><a href="read_blast.html">read_blast()</a></code> </p>
        </td>
        <td><p>Read a blast hit/alignment file.</p></td>
      </tr><tr><td>
          <p><code><a href="reads_to_phyloseq.html">reads_to_phyloseq()</a></code> </p>
        </td>
        <td><p>Converts read count data to a phyloseq object.</p></td>
      </tr><tr><td>
          <p><code><a href="remove_chimeras.html">remove_chimeras()</a></code> </p>
        </td>
        <td><p>Removes chimeric reads from amplicon sequencing data.</p></td>
      </tr><tr><td>
          <p><code><a href="remove_reference.html">remove_reference()</a></code> </p>
        </td>
        <td><p>Removes sequence that map to a given reference.</p></td>
      </tr><tr><td>
          <p><code><a href="select.html">select()</a></code> </p>
        </td>
        <td><p>Very simple select for miso artifacts. Just gets a particular entry.</p></td>
      </tr><tr><td>
          <p><code><a href="silva_build_taxonomy.html">silva_build_taxonomy()</a></code> </p>
        </td>
        <td><p>Build the lineages for the entire SILVA database.</p></td>
      </tr><tr><td>
          <p><code><a href="sra_submission.html">sra_submission()</a></code> </p>
        </td>
        <td><p>Prepare submission files for the NCBI sequence read archive (SRA).</p></td>
      </tr><tr><td>
          <p><code><a href="standardize.html">standardize()</a></code> </p>
        </td>
        <td><p>Standardize all continuous columns of a data frame.</p></td>
      </tr><tr><td>
          <p><code><a href="taxa_count.html">taxa_count()</a></code> </p>
        </td>
        <td><p>Counts the reads for a specific taxonomy level.</p></td>
      </tr><tr><td>
          <p><code><a href="taxa_metrics.html">taxa_metrics()</a></code> </p>
        </td>
        <td><p>Calculates what percentage of taxa was found in a reference set.</p></td>
      </tr><tr><td>
          <p><code><a href="taxa_quants.html">taxa_quants()</a></code> </p>
        </td>
        <td><p>Compares taxa quantification from a measurement to a reference ground truth.</p></td>
      </tr><tr><td>
          <p><code><a href="types.html">types()</a></code> </p>
        </td>
        <td><p>Applies the specified types to a data frame-like object.</p></td>
      </tr><tr><td>
          <p><code><a href="with_classification.html">with_classification()</a></code> </p>
        </td>
        <td><p>Quantify classification rates for each taxonomic rank.</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Christian Diener.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

